{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      Unnamed: 0  Database    Gene Amino acid position Reference Amino acid  \\\n",
      "0              0    Gnomad   GRIN1                   1                  Met   \n",
      "1              1    Gnomad   GRIN1                   3                  Thr   \n",
      "2              2    Gnomad   GRIN1                 117                  Pro   \n",
      "3              3    Gnomad   GRIN1                 243                  Ser   \n",
      "4              4    Gnomad   GRIN1                 288                  Val   \n",
      "...          ...       ...     ...                 ...                  ...   \n",
      "4582        4582  BCN-GRIN  GRIN2B                 732                  Asp   \n",
      "4583        4583  BCN-GRIN  GRIN2B                 737                  Glu   \n",
      "4584        4584  BCN-GRIN  GRIN2B                 832                  Ser   \n",
      "4585        4585  BCN-GRIN  GRIN2B                 821                  Val   \n",
      "4586        4586  BCN-GRIN  GRIN2B                 528                  Pro   \n",
      "\n",
      "     Mutated Amino acid  Equivalent position  \\\n",
      "0                     ?                    1   \n",
      "1                   Ala                    3   \n",
      "2                   Ala                  122   \n",
      "3                   Ala                  252   \n",
      "4                   Ala                  300   \n",
      "...                 ...                  ...   \n",
      "4582               Asn                   732   \n",
      "4583                Lys                  737   \n",
      "4584                Arg                  832   \n",
      "4585         GlyfsTer90                  821   \n",
      "4586                Thr                  528   \n",
      "\n",
      "                      Aminoacid change        Nucleotide change Domain  ...  \\\n",
      "0                NP_015566.1:p.(Met1?)       NM_007327.4:c.1A>T    ATD  ...   \n",
      "1              NP_015566.1:p.(Thr3Ala)       NM_007327.4:c.7A>G    ATD  ...   \n",
      "2            NP_015566.1:p.(Pro117Ala)     NM_007327.4:c.349C>G    ATD  ...   \n",
      "3            NP_015566.1:p.(Ser243Ala)  NM_001185090.1:c.790T>G    ATD  ...   \n",
      "4            NP_015566.1:p.(Val288Ala)     NM_007327.4:c.926T>C    ATD  ...   \n",
      "...                                ...                      ...    ...  ...   \n",
      "4582        NP_000825.2:p.(Asp732Asn )                c.2194G>A    LBD  ...   \n",
      "4583         NP_000825.2:p.(Glu737Lys)                c.2209G>A    LBD  ...   \n",
      "4584         NP_000825.2:p.(Ser832Arg)                c.2496C>G     M4  ...   \n",
      "4585  NP_000825.2:p.(Val821GlyfsTer90)                      NaN     M4  ...   \n",
      "4586         NP_000825.2:p.(Pro528Thr)                      NaN    LBD  ...   \n",
      "\n",
      "          NP_code Pubmed    Code          Ensembl VariationID AlleleID(s)  \\\n",
      "0     NP_015566.1    NaN  Q05586  ENSG00000176884         NaN         NaN   \n",
      "1     NP_015566.1    NaN  Q05586  ENSG00000176884         NaN         NaN   \n",
      "2     NP_015566.1    NaN  Q05586  ENSG00000176884         NaN         NaN   \n",
      "3     NP_015566.1    NaN  Q05586  ENSG00000176884         NaN         NaN   \n",
      "4     NP_015566.1    NaN  Q05586  ENSG00000176884         NaN         NaN   \n",
      "...           ...    ...     ...              ...         ...         ...   \n",
      "4582  NP_000825.2    NaN  Q13224  ENSG00000273079         NaN         NaN   \n",
      "4583  NP_000825.2    NaN  Q13224  ENSG00000273079         NaN         NaN   \n",
      "4584  NP_000825.2    NaN  Q13224  ENSG00000273079         NaN         NaN   \n",
      "4585  NP_000825.2    NaN  Q13224  ENSG00000273079         NaN         NaN   \n",
      "4586  NP_000825.2    NaN  Q13224  ENSG00000273079         NaN         NaN   \n",
      "\n",
      "     GRCh37Chromosome GRCh37Location GRCh38Chromosome GRCh38Location  \n",
      "0                 9.0      140033939              9.0      140033939  \n",
      "1                 9.0      140033945              9.0      140033945  \n",
      "2                 9.0      140036555              9.0      140036555  \n",
      "3                 9.0      140051176              9.0      140051176  \n",
      "4                 9.0      140051384              9.0      140051384  \n",
      "...               ...            ...              ...            ...  \n",
      "4582              NaN            NaN              NaN            NaN  \n",
      "4583              NaN            NaN              NaN            NaN  \n",
      "4584              NaN            NaN              NaN            NaN  \n",
      "4585              NaN            NaN              NaN            NaN  \n",
      "4586              NaN            NaN              NaN            NaN  \n",
      "\n",
      "[4587 rows x 27 columns]\n"
     ]
    }
   ],
   "source": [
    "#Llegeix taula\n",
    "table = pd.read_excel('GRINDB_2023-04-18.xlsx')\n",
    "print(table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        Gene Amino acid position Reference Amino acid Mutated Amino acid  \\\n",
      "0      GRIN1                   1                  Met                  ?   \n",
      "1      GRIN1                   3                  Thr                Ala   \n",
      "2      GRIN1                 117                  Pro                Ala   \n",
      "3      GRIN1                 243                  Ser                Ala   \n",
      "4      GRIN1                 288                  Val                Ala   \n",
      "...      ...                 ...                  ...                ...   \n",
      "4582  GRIN2B                 732                  Asp               Asn    \n",
      "4583  GRIN2B                 737                  Glu                Lys   \n",
      "4584  GRIN2B                 832                  Ser                Arg   \n",
      "4585  GRIN2B                 821                  Val         GlyfsTer90   \n",
      "4586  GRIN2B                 528                  Pro                Thr   \n",
      "\n",
      "     Gain/Loss Disease-association Mutation Category  \n",
      "0          NaN             Neutral          Missense  \n",
      "1          NaN             Neutral          Missense  \n",
      "2          NaN             Neutral          Missense  \n",
      "3          NaN             Neutral          Missense  \n",
      "4          NaN             Neutral          Missense  \n",
      "...        ...                 ...               ...  \n",
      "4582       NaN  Disease associated          Missense  \n",
      "4583       NaN  Disease associated          Missense  \n",
      "4584       NaN  Disease associated          Missense  \n",
      "4585       NaN  Disease associated        Truncation  \n",
      "4586       NaN  Disease associated          Missense  \n",
      "\n",
      "[4587 rows x 7 columns]\n"
     ]
    }
   ],
   "source": [
    "#Eliminem columnes que no ens interessen\n",
    "#Ens qudem només amb: Gene, Amino acid position, Gain/Loss, Disease-association, Mutation Category, Reference Amino acid, Mutated Amino acid\n",
    "taula2=table.drop(['Unnamed: 0', 'Database','Equivalent position','Aminoacid change','Nucleotide change','Domain','Functional Information','Phenotype category','Clinical description','NM_code','NP_code','Pubmed','Code','Ensembl','VariationID','AlleleID(s)','GRCh37Chromosome','GRCh37Location','GRCh38Chromosome','GRCh38Location'], axis=1)\n",
    "print(taula2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Per GRIN1\n",
    "#Agafem de la taula només GRIN1 i creem nova taula\n",
    "G1=taula2.loc[taula2['Gene']=='GRIN1']\n",
    "#De la nova taula agafem només Missense i creem una altra nova taula\n",
    "G1_1=G1.loc[G1['Mutation Category']=='Missense']\n",
    "#De la nova taula agafem només les Disease associated, Uncertain pathogenesis, i neutral creem una altra noves taules\n",
    "#Taula Disease associated\n",
    "G1_da=G1_1.loc[G1['Disease-association']=='Disease associated']\n",
    "#Taula Uncertain pathogenesis\n",
    "G1_up=G1_1.loc[G1['Disease-association']=='Uncertain pathogenesis']\n",
    "#Taula Neutral\n",
    "G1_n=G1_1.loc[G1['Disease-association']=='Neutral']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'resi 110 or resi 118 or resi 217 or resi 217 or resi 227 or resi 248 or resi 370 or resi 397 or resi 523 or resi 531 or resi 532 or resi 532 or resi 536 or resi 548 or resi 549 or resi 549 or resi 549 or resi 551 or resi 552 or resi 556 or resi 557 or resi 557 or resi 559 or resi 560 or resi 597 or resi 616 or resi 617 or resi 618 or resi 620 or resi 637 or resi 637 or resi 638 or resi 638 or resi 641 or resi 641 or resi 642 or resi 644 or resi 645 or resi 647 or resi 647 or resi 650 or resi 650 or resi 650 or resi 652 or resi 653 or resi 654 or resi 656 or resi 659 or resi 659 or resi 660 or resi 662 or resi 670 or resi 674 or resi 682 or resi 688 or resi 688 or resi 706 or resi 732 or resi 737 or resi 744 or resi 766 or resi 789 or resi 794 or resi 794 or resi 803 or resi 805 or resi 805 or resi 806 or resi 806 or resi 813 or resi 814 or resi 815 or resi 815 or resi 816 or resi 817 or resi 827 or resi 830 or resi 834 or resi 835 or resi 835 or resi 844 or resi 844 or resi 844 or resi 844 or resi 851'"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#GRIN 1 LLISTA DISEASE ASSOCIATED\n",
    "#creo una llista amb la posició del aminoàcid\n",
    "llista1=list(G1_da['Amino acid position'])\n",
    "#la ordeno per ordre creixent\n",
    "llista1.sort()\n",
    "#creo llista buida\n",
    "L1=[]\n",
    "#inicialitzo comptador a 0\n",
    "comp=0\n",
    "#mentres que la llargada de la llista sigui diferent al comptador\n",
    "while len(llista1)!=comp:\n",
    "    #poso que \"h\" és resi, la posició de la llista de aa, i or\n",
    "    h=['resi',llista1[comp],'or ']\n",
    "    #afegeixo h a la llista buida L1\n",
    "    L1.append(h)\n",
    "    #sumo 1 al comptador per canviar de posició\n",
    "    comp=comp+1\n",
    "#paso la llista a cadena\n",
    "s = ','.join(map(str, L1))\n",
    "#elimino: [ ] , '\n",
    "ss=s.replace('[','')\n",
    "sss=ss.replace(',','')\n",
    "ssss=sss.replace(']','')\n",
    "sssss=ssss.replace(\"'\",\"\")\n",
    "#ara borrem el \"or\" final \n",
    "#faig una funció per borrar el or final\n",
    "def remove_last_or(chain):\n",
    "    #Divideix la cadena en diferents elements separats per \"or\"\n",
    "    elements = chain.split(\"or\")\n",
    "    #Per si nomes hagues un a la llista, retorna la cadena tal com està\n",
    "    if len(elements) == 1:\n",
    "        return chain\n",
    "    #Treu qualsevol espai en blanc inicial o final de cada element\n",
    "    cleaned_elements = [elem.strip() for elem in elements]\n",
    "    #Elimina l'últim element, que hauria de ser una cadena buida després de l'últim \"or\"\n",
    "    cleaned_elements.pop()\n",
    "    #Torna a unir els elements \"netejats\" fent servir \"or\" com a separador\n",
    "    new_chain = \" or \".join(cleaned_elements)\n",
    "    return new_chain\n",
    "Grin1_da_aa = remove_last_or(sssss)\n",
    "Grin1_da_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'resi 27 or resi 117 or resi 123 or resi 217 or resi 306 or resi 349 or resi 349 or resi 370 or resi 419 or resi 540 or resi 548 or resi 641 or resi 643 or resi 655 or resi 658 or resi 664 or resi 670 or resi 798 or resi 807 or resi 818 or resi 855 or resi 939'"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#GRIN 1 LLISTA Uncertain pathogenesis\n",
    "#creo una llista amb la posició del aminoàcid\n",
    "llista1=list(G1_up['Amino acid position'])\n",
    "#la ordeno per ordre creixent\n",
    "llista1.sort()\n",
    "#creo llista buida\n",
    "L1=[]\n",
    "#inicialitzo comptador a 0\n",
    "comp=0\n",
    "#mentres que la llargada de la llista sigui diferent al comptador\n",
    "while len(llista1)!=comp:\n",
    "    #poso que \"h\" és resi, la posició de la llista de aa, i or\n",
    "    h=['resi',llista1[comp],'or ']\n",
    "    #afegeixo h a la llista buida L1\n",
    "    L1.append(h)\n",
    "    #sumo 1 al comptador per canviar de posició\n",
    "    comp=comp+1\n",
    "#paso la llista a cadena\n",
    "s = ','.join(map(str, L1))\n",
    "#elimino: [ ] , '\n",
    "ss=s.replace('[','')\n",
    "sss=ss.replace(',','')\n",
    "ssss=sss.replace(']','')\n",
    "sssss=ssss.replace(\"'\",\"\")\n",
    "#ara borrem el \"or\" final \n",
    "Grin1_up_aa = remove_last_or(sssss)\n",
    "Grin1_up_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'resi 1 or resi 1 or resi 2 or resi 3 or resi 5 or resi 5 or resi 5 or resi 8 or resi 8 or resi 20 or resi 24 or resi 27 or resi 28 or resi 33 or resi 38 or resi 39 or resi 40 or resi 46 or resi 49 or resi 51 or resi 52 or resi 52 or resi 68 or resi 70 or resi 70 or resi 71 or resi 75 or resi 80 or resi 81 or resi 98 or resi 98 or resi 99 or resi 102 or resi 115 or resi 115 or resi 117 or resi 124 or resi 124 or resi 139 or resi 139 or resi 142 or resi 149 or resi 153 or resi 155 or resi 156 or resi 156 or resi 158 or resi 169 or resi 169 or resi 170 or resi 170 or resi 174 or resi 174 or resi 175 or resi 179 or resi 196 or resi 198 or resi 198 or resi 201 or resi 204 or resi 205 or resi 206 or resi 209 or resi 210 or resi 211 or resi 217 or resi 219 or resi 225 or resi 231 or resi 234 or resi 235 or resi 243 or resi 249 or resi 252 or resi 257 or resi 258 or resi 267 or resi 268 or resi 280 or resi 282 or resi 284 or resi 287 or resi 288 or resi 290 or resi 301 or resi 304 or resi 307 or resi 316 or resi 326 or resi 331 or resi 331 or resi 337 or resi 338 or resi 341 or resi 344 or resi 346 or resi 350 or resi 351 or resi 351 or resi 353 or resi 359 or resi 362 or resi 364 or resi 365 or resi 366 or resi 370 or resi 374 or resi 384 or resi 390 or resi 390 or resi 394 or resi 411 or resi 416 or resi 424 or resi 424 or resi 428 or resi 429 or resi 430 or resi 433 or resi 435 or resi 437 or resi 440 or resi 442 or resi 444 or resi 444 or resi 445 or resi 446 or resi 447 or resi 447 or resi 447 or resi 447 or resi 448 or resi 450 or resi 453 or resi 457 or resi 461 or resi 461 or resi 465 or resi 471 or resi 495 or resi 507 or resi 507 or resi 512 or resi 515 or resi 522 or resi 524 or resi 539 or resi 547 or resi 582 or resi 583 or resi 589 or resi 591 or resi 593 or resi 629 or resi 652 or resi 685 or resi 694 or resi 709 or resi 741 or resi 746 or resi 765 or resi 767 or resi 772 or resi 772 or resi 772 or resi 796 or resi 818 or resi 825 or resi 833 or resi 839 or resi 839 or resi 845 or resi 849 or resi 853 or resi 853 or resi 854 or resi 855 or resi 856 or resi 856 or resi 856 or resi 857 or resi 859 or resi 860 or resi 860 or resi 864 or resi 865 or resi 865 or resi 866 or resi 867 or resi 867 or resi 868 or resi 870 or resi 870 or resi 873 or resi 875 or resi 875 or resi 877 or resi 877 or resi 877 or resi 877 or resi 877 or resi 879 or resi 880 or resi 880 or resi 881 or resi 882 or resi 882 or resi 882 or resi 883 or resi 883 or resi 884 or resi 886 or resi 886 or resi 886 or resi 896 or resi 897 or resi 898 or resi 898 or resi 900 or resi 900 or resi 901 or resi 901 or resi 901 or resi 902 or resi 903 or resi 904 or resi 906 or resi 907 or resi 907 or resi 908 or resi 911 or resi 911 or resi 919 or resi 919 or resi 921 or resi 928'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#GRIN 1 LLISTA Neutral\n",
    "#creo una llista amb la posició del aminoàcid\n",
    "llista1=list(G1_n['Amino acid position'])\n",
    "#la ordeno per ordre creixent\n",
    "llista1.sort()\n",
    "#creo llista buida\n",
    "L1=[]\n",
    "#inicialitzo comptador a 0\n",
    "comp=0\n",
    "#mentres que la llargada de la llista sigui diferent al comptador\n",
    "while len(llista1)!=comp:\n",
    "    #poso que \"h\" és resi, la posició de la llista de aa, i or\n",
    "    h=['resi',llista1[comp],'or ']\n",
    "    #afegeixo h a la llista buida L1\n",
    "    L1.append(h)\n",
    "    #sumo 1 al comptador per canviar de posició\n",
    "    comp=comp+1\n",
    "#paso la llista a cadena\n",
    "s = ','.join(map(str, L1))\n",
    "#elimino: [ ] , '\n",
    "ss=s.replace('[','')\n",
    "sss=ss.replace(',','')\n",
    "ssss=sss.replace(']','')\n",
    "sssss=ssss.replace(\"'\",\"\")\n",
    "#ara borrem el \"or\" final \n",
    "Grin1_n_aa = remove_last_or(sssss)\n",
    "Grin1_n_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "llista dels aa que estan a gnomAD V2 i no a GRIN db:  [['Arg', '36', 'Gln'], ['Val', '141', 'Met'], ['Glu', '153', 'Asp'], ['Arg', '156', 'Leu'], ['Arg', '179', 'His'], ['Arg', '187', 'His'], ['Asp', '200', 'Asn'], ['Arg', '208', 'Cys'], ['Glu', '213', 'Gly'], ['Gln', '217', 'Leu'], ['Gln', '217', 'Arg'], ['Asp', '219', 'Asn'], ['Asp', '219', 'Val'], ['Thr', '222', 'Ile'], ['Val', '225', 'Met'], ['Thr', '226', 'Met'], ['Ala', '227', 'Val'], ['Leu', '229', 'Met'], ['Met', '230', 'Thr'], ['Glu', '231', 'Asp'], ['Ala', '232', 'Thr'], ['Ala', '232', 'Val'], ['Arg', '238', 'Gln'], ['Ile', '240', 'Val'], ['Ile', '240', 'Phe'], ['Glu', '246', 'Gly'], ['Asp', '248', 'Asn'], ['Val', '252', 'Ile'], ['Ala', '255', 'Thr'], ['Ala', '256', 'Pro'], ['Ala', '257', 'Thr'], ['Thr', '262', 'Arg'], ['Gly', '263', 'Ala'], ['Ser', '264', 'Ala'], ['Tyr', '266', 'Phe'], ['Val', '270', 'Leu'], ['Gly', '271', 'Asp'], ['Arg', '273', 'Leu'], ['Glu', '274', 'Lys'], ['Asn', '278', 'Ser'], ['Ala', '279', 'Asp'], ['Leu', '288', 'Ile'], ['Gly', '289', 'Arg'], ['His', '301', 'Pro'], ['Ser', '303', 'Asn'], ['Ala', '305', 'Pro'], ['Val', '308', 'Leu'], ['Val', '309', 'Ala'], ['Gln', '311', 'His'], ['Ala', '312', 'Val'], ['Ile', '322', 'Ser'], ['Pro', '325', 'Arg'], ['Gly', '328', 'Val'], ['Asn', '334', 'Ser'], ['Trp', '336', 'Arg'], ['Met', '347', 'Val'], ['Ala', '352', 'Val'], ['Ala', '352', 'Glu'], ['Arg', '358', 'His'], ['Val', '359', 'Met'], ['Asn', '362', 'Ser'], ['Arg', '367', 'Gln'], ['Ala', '370', 'Ser'], ['Ala', '370', 'Thr'], ['Asn', '371', 'Ser'], ['Tyr', '372', 'His'], ['Tyr', '372', 'Cys'], ['Ile', '374', 'Val'], ['Arg', '380', 'His'], ['Val', '383', 'Met'], ['Val', '385', 'Met'], ['Gly', '386', 'Arg'], ['Ile', '387', 'Met'], ['Thr', '391', 'Ser'], ['Pro', '395', 'Leu'], ['Gly', '405', 'Arg'], ['Arg', '411', 'Gly'], ['Arg', '411', 'Gln'], ['Met', '415', 'Ile'], ['Val', '432', 'Ile'], ['Ser', '437', 'Arg'], ['Thr', '440', 'Ala'], ['Cys', '441', 'Arg'], ['Phe', '445', 'Leu'], ['Phe', '445', 'Ser'], ['Gly', '449', 'Arg'], ['Asp', '450', 'Asn'], ['Pro', '451', 'Ser'], ['Lys', '454', 'Thr'], ['Ile', '456', 'Asn'], ['Thr', '458', 'Ala'], ['Asn', '461', 'Thr'], ['Pro', '465', 'Arg'], ['Pro', '465', 'Leu'], ['Gly', '466', 'Ser'], ['Gly', '466', 'Ala'], ['Ser', '467', 'Asn'], ['Pro', '468', 'Ala'], ['Pro', '468', 'Ser'], ['Pro', '468', 'His'], ['Pro', '468', 'Leu'], ['Arg', '469', 'His'], ['Thr', '471', 'Ala'], ['Gln', '474', 'His'], ['Gly', '478', 'Ser'], ['Asp', '482', 'Val'], ['Asp', '482', 'Gly'], ['Lys', '486', 'Met'], ['Asn', '492', 'Thr'], ['Lys', '516', 'Asn'], ['Met', '522', 'Arg'], ['Leu', '527', 'Phe'], ['Ser', '528', 'Asn'], ['Ser', '528', 'Arg'], ['Met', '533', 'Leu'], ['Met', '533', 'Thr'], ['Ala', '536', 'Val'], ['Leu', '538', 'Gln'], ['Ile', '540', 'Lys'], ['Glu', '543', 'Gln'], ['Glu', '543', 'Asp'], ['Ala', '545', 'Pro'], ['Lys', '565', 'Arg'], ['Lys', '565', 'Asn'], ['Pro', '568', 'Thr'], ['Pro', '568', 'Leu'], ['Arg', '569', 'Trp'], ['Trp', '584', 'Leu'], ['Val', '596', 'Met'], ['Arg', '603', 'His'], ['Phe', '604', 'Leu'], ['Ser', '605', 'Thr'], ['Ser', '614', 'Ile'], ['Ser', '614', 'Thr'], ['Ala', '650', 'Val'], ['Gly', '659', 'Asp'], ['Arg', '680', 'Trp'], ['Pro', '681', 'Leu'], ['Gly', '687', 'Ser'], ['Arg', '692', 'Trp'], ['Lys', '706', 'Arg'], ['Arg', '715', 'Pro'], ['Glu', '728', 'Lys'], ['His', '730', 'Tyr'], ['Ala', '735', 'Val'], ['Asn', '745', 'Tyr'], ['Ser', '762', 'Trp'], ['Leu', '767', 'Pro'], ['Gly', '780', 'Asp'], ['Arg', '784', 'Cys'], ['Asp', '786', 'Gly'], ['Pro', '788', 'Leu'], ['Val', '793', 'Leu'], ['Val', '793', 'Ile'], ['Val', '793', 'Ala'], ['Gln', '817', 'His'], ['Val', '846', 'Met'], ['Gly', '848', 'Arg'], ['Ile', '854', 'Leu'], ['Arg', '860', 'Gly'], ['Arg', '860', 'Gln'], ['Arg', '866', 'Gln'], ['Val', '878', 'Leu'], ['Pro', '874', 'Ser'], ['Pro', '874', 'Leu'], ['Arg', '875', 'Gly'], ['Ala', '876', 'Ser'], ['Gly', '881', 'Asp'], ['Ser', '885', 'Phe'], ['Pro', '886', 'Arg'], ['Pro', '886', 'Leu'], ['Lys', '887', 'Arg'], ['Ala', '888', 'Ser'], ['Ala', '888', 'Thr'], ['Val', '889', 'Met'], ['Ala', '894', 'Thr'], ['Val', '896', 'Ala'], ['Gly', '900', 'Arg'], ['Ser', '903', 'Asn'], ['Ser', '903', 'Thr'], ['His', '904', 'Gln'], ['Gly', '905', 'Val'], ['Gly', '907', 'Glu'], ['Ala', '891', 'Thr'], ['Thr', '907', 'Ile'], ['Ser', '918', 'Phe'], ['Thr', '921', 'Met'], ['Ser', '922', 'Arg'], ['Arg', '927', 'Cys'], ['Arg', '927', 'His'], ['Val', '937', 'Glu'], ['Pro', '939', 'Leu'], ['Arg', '940', 'Gln'], ['Ile', '943', 'Thr'], ['Gln', '922', 'Arg'], ['Tyr', '923', 'Cys'], ['Asp', '927', 'Tyr'], ['Ile', '928', 'Thr'], ['Thr', '929', 'Met'], ['Leu', '932', 'Val'], ['Leu', '932', 'Pro'], ['Ser', '940', 'Gly'], ['Val', '942', 'Leu']]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ys/tvdvp6kj24df4dh9qb_87_tr0000gn/T/ipykernel_43013/2974005338.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2=G1_rev[['lletres', 'numeros']] = G1_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n",
      "/var/folders/ys/tvdvp6kj24df4dh9qb_87_tr0000gn/T/ipykernel_43013/2974005338.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2=G1_rev[['lletres', 'numeros']] = G1_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n"
     ]
    }
   ],
   "source": [
    "#COMPARACIÓ GRIN1 GRINDB VS GNOMAD V2\n",
    "\n",
    "#De les que tinc a la base de dades\n",
    "#Per fer Met654Ile de les neutral\n",
    "#copio la llista de grin1 neutral\n",
    "G1_aa=G1_n\n",
    "#alimino totes les columnes que no necessito\n",
    "G1_aa=G1_aa.drop(['Gene','Gain/Loss','Disease-association','Mutation Category'], axis=1)\n",
    "#les reordeno\n",
    "G1_aa = G1_aa.reindex(columns=['Reference Amino acid','Amino acid position','Mutated Amino acid'])\n",
    "#Poso com string la posició del aa mutat\n",
    "G1_aa['string_numbers'] = G1_aa['Amino acid position'].astype(str)\n",
    "#alimino totes les columnes que no necessito\n",
    "G1_aa=G1_aa.drop(['Amino acid position'], axis=1)\n",
    "#les reordeno\n",
    "G1_aa = G1_aa.reindex(columns=['Reference Amino acid','string_numbers','Mutated Amino acid'])\n",
    "#ho paso a llista\n",
    "G1_aaL=G1_aa.to_numpy().tolist()\n",
    "#ara tinc la llista dels aa mutació neutra de grin1 que estan a la base de dades\n",
    "G1_aaL\n",
    "\n",
    "\n",
    "#llegeixo arxiu gnomad v2\n",
    "dades = pd.read_csv(\"/Users/ismaelvera/Desktop/TFG/Mutacions GRINdb/GRIN_missense_variants_gnomad_v2_1.txt\",header=None) \n",
    "#em quedo nomes amb les columnes que diuen la mutació i el gen\n",
    "dades=dades.drop([0,1,2,3,4,5,6,8,10,11,12,13,14], axis=1)\n",
    "#faig una llista amb els que son nomes de grin1\n",
    "G1_rev=dades.loc[dades[9]=='GRIN1']\n",
    "#com que ho tinc com a p.Ser2Arg\n",
    "#creo funció per separar lletres de numeros \n",
    "def separate_letters_numbers(text):\n",
    "    lletres = ''.join(filter(str.isalpha, text))\n",
    "    numeros = ''.join(filter(str.isdigit, text))\n",
    "    return (lletres, numeros)\n",
    "#aplico funció a la columna que m'interesa\n",
    "df2=G1_rev[['lletres', 'numeros']] = G1_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n",
    "#ara tinc lletres i numeros separats\n",
    "#separo les lletres per tenir p, Ser, Arg\n",
    "df2['col1'] = df2[0].str[0]\n",
    "df2['col2'] = df2[0].str[1:4]\n",
    "df2['col3'] = df2[0].str[4:]\n",
    "#elimino totes les columnes que no necessito\n",
    "df2=df2.drop([0,'col1'], axis=1)\n",
    "#les reordeno\n",
    "df2 = df2.reindex(columns=['col2',1,'col3'])\n",
    "#ho paso a llista\n",
    "df2L1=df2.to_numpy().tolist()\n",
    "\n",
    "\n",
    "#faig la comparació\n",
    "#he de crear una llista buida on guardare els que no estan \n",
    "noestanG1V2=[]\n",
    "#per cada element que tinc a la llista de gnomad v2, si no esta a la llista dels \n",
    "# que tinc a la base de dades, mel afegeix a la llista L1\n",
    "for element in df2L1:\n",
    "    if element not in G1_aaL:\n",
    "        noestanG1V2.append(element)\n",
    "\n",
    "\n",
    "#ara per poder marcar-los al model molecular he de fer que posi resi 23 or resi 234 ....\n",
    "#primer separo la llista L1, que son els que no tinc en AA1 posició AA2\n",
    "df1 = pd.DataFrame(noestanG1V2, columns=['AA 1', 'Position', 'AA 2'])\n",
    "#creo una llista amb la posició del aminoàcid\n",
    "llista1=list(df1['Position'])\n",
    "#la ordeno per ordre creixent\n",
    "llista1.sort()\n",
    "#creo llista buida\n",
    "L1=[]\n",
    "#inicialitzo comptador a 0\n",
    "comp=0\n",
    "#mentres que la llargada de la llista sigui diferent al comptador\n",
    "while len(llista1)!=comp:\n",
    "    #poso que \"h\" és resi, la posició de la llista de aa, i or\n",
    "    h=['resi',llista1[comp],'or ']\n",
    "    #afegeixo h a la llista buida L1\n",
    "    L1.append(h)\n",
    "    #sumo 1 al comptador per canviar de posició\n",
    "    comp=comp+1\n",
    "#paso la llista a cadena\n",
    "s = ','.join(map(str, L1))\n",
    "#elimino: [ ] , '\n",
    "ss=s.replace('[','')\n",
    "sss=ss.replace(',','')\n",
    "ssss=sss.replace(']','')\n",
    "sssss=ssss.replace(\"'\",\"\")\n",
    "#ara borrem el \"or\" final \n",
    "#faig una funció per borrar el or final\n",
    "def remove_last_or(chain):\n",
    "    #Divideix la cadena en diferents elements separats per \"or\"\n",
    "    elements = chain.split(\"or\")\n",
    "    #Per si nomes hagues un a la llista, retorna la cadena tal com està\n",
    "    if len(elements) == 1:\n",
    "        return chain\n",
    "    #Treu qualsevol espai en blanc inicial o final de cada element\n",
    "    cleaned_elements = [elem.strip() for elem in elements]\n",
    "    #Elimina l'últim element, que hauria de ser una cadena buida després de l'últim \"or\"\n",
    "    cleaned_elements.pop()\n",
    "    #Torna a unir els elements \"netejats\" fent servir \"or\" com a separador\n",
    "    new_chain = \" or \".join(cleaned_elements)\n",
    "    return new_chain\n",
    "G1_rev_v2 = remove_last_or(sssss)\n",
    "G1_rev_v2\n",
    "\n",
    "#mostro la llista dels que no estan\n",
    "print('llista dels aa que estan a gnomAD V2 i no a GRIN db: ',noestanG1V2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "llista dels aa que estan a gnomAD V2 i no a GRIN db:  [['Ala', '18', 'Gly'], ['Pro', '24', 'Arg'], ['Arg', '36', 'Gln'], ['Met', '41', 'Ile'], ['Lys', '57', 'Thr'], ['Ile', '58', 'Phe'], ['Ser', '84', 'Phe'], ['Pro', '96', 'Ser'], ['Gly', '112', 'Ser'], ['Arg', '115', 'Gly'], ['Thr', '123', 'Asn'], ['Pro', '142', 'Gln'], ['Arg', '156', 'Ser'], ['Tyr', '196', 'Cys'], ['Arg', '208', 'Cys'], ['Glu', '192', 'Gly'], ['Glu', '192', 'Asp'], ['Asp', '198', 'Glu'], ['Ala', '211', 'Val'], ['Val', '231', 'Leu'], ['Met', '240', 'Lys'], ['Gln', '290', 'Arg'], ['His', '293', 'Gln'], ['Asp', '303', 'Asn'], ['Pro', '304', 'Ser'], ['Asn', '313', 'Ser'], ['Trp', '315', 'Arg'], ['Gly', '318', 'Arg'], ['Asp', '332', 'Glu'], ['Val', '334', 'Met'], ['Ala', '349', 'Ser'], ['Ala', '349', 'Thr'], ['Asn', '368', 'Lys'], ['Thr', '370', 'Ala'], ['Val', '372', 'Gly'], ['Lys', '399', 'Glu'], ['Ser', '416', 'Asn'], ['Thr', '419', 'Ala'], ['Thr', '425', 'Pro'], ['Gly', '438', 'Arg'], ['Asn', '440', 'Lys'], ['Gly', '445', 'Asp'], ['Arg', '448', 'Ser'], ['Trp', '498', 'Gly'], ['Gly', '503', 'Ser'], ['Gly', '503', 'Asp'], ['Leu', '506', 'Phe'], ['Val', '514', 'Leu'], ['Val', '514', 'Leu'], ['Arg', '523', 'Leu'], ['Phe', '529', 'Val'], ['Phe', '529', 'Ser'], ['Leu', '541', 'Val'], ['Leu', '541', 'Arg'], ['Val', '542', 'Ala'], ['Lys', '544', 'Arg'], ['Arg', '548', 'Trp'], ['Ser', '593', 'Ile'], ['Ala', '600', 'Thr'], ['Ala', '715', 'Val'], ['Asn', '724', 'Tyr'], ['Pro', '767', 'Ser'], ['Arg', '794', 'Trp'], ['Leu', '862', 'Met'], ['Arg', '894', 'His'], ['Ser', '901', 'Asn'], ['Arg', '906', 'Ser'], ['Glu', '926', 'Gly'], ['Gln', '922', 'Arg'], ['Tyr', '923', 'Cys'], ['Ile', '928', 'Thr'], ['Ser', '938', 'Trp'], ['Ser', '940', 'Gly'], ['Val', '942', 'Leu']]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ys/tvdvp6kj24df4dh9qb_87_tr0000gn/T/ipykernel_43013/622243419.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2=G1_rev[['lletres', 'numeros']] = G1_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n",
      "/var/folders/ys/tvdvp6kj24df4dh9qb_87_tr0000gn/T/ipykernel_43013/622243419.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2=G1_rev[['lletres', 'numeros']] = G1_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n"
     ]
    }
   ],
   "source": [
    "#COMPARACIÓ GRIN1 GRINDB VS GNOMAD V3\n",
    "\n",
    "#De les que tinc a la base de dades\n",
    "#Per fer Met654Ile de les neutral\n",
    "#copio la llista de grin1 neutral\n",
    "G1_aa=G1_n\n",
    "#alimino totes les columnes que no necessito\n",
    "G1_aa=G1_aa.drop(['Gene','Gain/Loss','Disease-association','Mutation Category'], axis=1)\n",
    "#les reordeno\n",
    "G1_aa = G1_aa.reindex(columns=['Reference Amino acid','Amino acid position','Mutated Amino acid'])\n",
    "#Poso com string la posició del aa mutat\n",
    "G1_aa['string_numbers'] = G1_aa['Amino acid position'].astype(str)\n",
    "#alimino totes les columnes que no necessito\n",
    "G1_aa=G1_aa.drop(['Amino acid position'], axis=1)\n",
    "#les reordeno\n",
    "G1_aa = G1_aa.reindex(columns=['Reference Amino acid','string_numbers','Mutated Amino acid'])\n",
    "#ho paso a llista\n",
    "G1_aaL=G1_aa.to_numpy().tolist()\n",
    "#ara tinc la llista dels aa mutació neutra de grin1 que estan a la base de dades\n",
    "G1_aaL\n",
    "\n",
    "\n",
    "#llegeixo arxiu gnomad v2\n",
    "dades = pd.read_csv(\"/Users/ismaelvera/Desktop/TFG/Mutacions GRINdb/GRIN_missense_variants_gnomad_v3.txt\",header=None) \n",
    "#em quedo nomes amb les columnes que diuen la mutació i el gen\n",
    "dades=dades.drop([0,1,2,3,4,5,6,8,10,11,12,13,14], axis=1)\n",
    "#faig una llista amb els que son nomes de grin1\n",
    "G1_rev=dades.loc[dades[9]=='GRIN1']\n",
    "#com que ho tinc com a p.Ser2Arg\n",
    "#creo funció per separar lletres de numeros \n",
    "def separate_letters_numbers(text):\n",
    "    lletres = ''.join(filter(str.isalpha, text))\n",
    "    numeros = ''.join(filter(str.isdigit, text))\n",
    "    return (lletres, numeros)\n",
    "#aplico funció a la columna que m'interesa\n",
    "df2=G1_rev[['lletres', 'numeros']] = G1_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n",
    "#ara tinc lletres i numeros separats\n",
    "#separo les lletres per tenir p, Ser, Arg\n",
    "df2['col1'] = df2[0].str[0]\n",
    "df2['col2'] = df2[0].str[1:4]\n",
    "df2['col3'] = df2[0].str[4:]\n",
    "#elimino totes les columnes que no necessito\n",
    "df2=df2.drop([0,'col1'], axis=1)\n",
    "#les reordeno\n",
    "df2 = df2.reindex(columns=['col2',1,'col3'])\n",
    "#ho paso a llista\n",
    "df2L1=df2.to_numpy().tolist()\n",
    "\n",
    "\n",
    "#faig la comparació\n",
    "#he de crear una llista buida on guardare els que no estan \n",
    "noestanG1V3=[]\n",
    "#per cada element que tinc a la llista de gnomad v2, si no esta a la llista dels \n",
    "# que tinc a la base de dades, mel afegeix a la llista L1\n",
    "for element in df2L1:\n",
    "    if element not in G1_aaL:\n",
    "        noestanG1V3.append(element)\n",
    "\n",
    "\n",
    "#ara per poder marcar-los al model molecular he de fer que posi resi 23 or resi 234 ....\n",
    "#primer separo la llista L1, que son els que no tinc en AA1 posició AA2\n",
    "df1 = pd.DataFrame(noestanG1V3, columns=['AA 1', 'Position', 'AA 2'])\n",
    "#creo una llista amb la posició del aminoàcid\n",
    "llista1=list(df1['Position'])\n",
    "#la ordeno per ordre creixent\n",
    "llista1.sort()\n",
    "#creo llista buida\n",
    "L1=[]\n",
    "#inicialitzo comptador a 0\n",
    "comp=0\n",
    "#mentres que la llargada de la llista sigui diferent al comptador\n",
    "while len(llista1)!=comp:\n",
    "    #poso que \"h\" és resi, la posició de la llista de aa, i or\n",
    "    h=['resi',llista1[comp],'or ']\n",
    "    #afegeixo h a la llista buida L1\n",
    "    L1.append(h)\n",
    "    #sumo 1 al comptador per canviar de posició\n",
    "    comp=comp+1\n",
    "#paso la llista a cadena\n",
    "s = ','.join(map(str, L1))\n",
    "#elimino: [ ] , '\n",
    "ss=s.replace('[','')\n",
    "sss=ss.replace(',','')\n",
    "ssss=sss.replace(']','')\n",
    "sssss=ssss.replace(\"'\",\"\")\n",
    "#ara borrem el \"or\" final \n",
    "#faig una funció per borrar el or final\n",
    "def remove_last_or(chain):\n",
    "    #Divideix la cadena en diferents elements separats per \"or\"\n",
    "    elements = chain.split(\"or\")\n",
    "    #Per si nomes hagues un a la llista, retorna la cadena tal com està\n",
    "    if len(elements) == 1:\n",
    "        return chain\n",
    "    #Treu qualsevol espai en blanc inicial o final de cada element\n",
    "    cleaned_elements = [elem.strip() for elem in elements]\n",
    "    #Elimina l'últim element, que hauria de ser una cadena buida després de l'últim \"or\"\n",
    "    cleaned_elements.pop()\n",
    "    #Torna a unir els elements \"netejats\" fent servir \"or\" com a separador\n",
    "    new_chain = \" or \".join(cleaned_elements)\n",
    "    return new_chain\n",
    "G1_rev_v3 = remove_last_or(sssss)\n",
    "G1_rev_v3\n",
    "\n",
    "#mostro la llista dels que no estan\n",
    "print('llista dels aa que estan a gnomAD V2 i no a GRIN db: ',noestanG1V3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Per GRIN2A\n",
    "#Agafem de la taula només GRIN1 i creem nova taula\n",
    "G2A=taula2.loc[taula2['Gene']=='GRIN2A']\n",
    "#De la nova taula agafem només Missense i creem una altra nova taula\n",
    "G2A_1=G2A.loc[G2A['Mutation Category']=='Missense']\n",
    "#De la nova taula agafem només les Disease associated, Uncertain pathogenesis, i neutral creem una altra noves taules\n",
    "#Taula Disease associated\n",
    "G2A_da=G2A_1.loc[G2A['Disease-association']=='Disease associated']\n",
    "#Taula Uncertain pathogenesis\n",
    "G2A_up=G2A_1.loc[G2A['Disease-association']=='Uncertain pathogenesis']\n",
    "#Taula Neutral\n",
    "G2A_n=G2A_1.loc[G2A['Disease-association']=='Neutral']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'resi 1 or resi 1 or resi 10 or resi 31 or resi 57 or resi 79 or resi 98 or resi 184 or resi 231 or resi 231 or resi 243 or resi 366 or resi 819 or resi 823 or resi nan or resi 252 or resi 270 or resi 278 or resi 295 or resi 349 or resi 371 or resi 373 or resi 380 or resi 411 or resi 436 or resi 436 or resi 443 or resi 449 or resi 456 or resi 458 or resi 459 or resi 483 or resi 490 or resi 498 or resi 500 or resi 511 or resi 513 or resi 518 or resi 518 or resi 518 or resi 527 or resi 531 or resi 548 or resi 548 or resi 551 or resi 552 or resi 554 or resi 558 or resi 563 or resi 576 or resi 598 or resi 611 or resi 614 or resi 615 or resi 615 or resi 632 or resi 635 or resi 639 or resi 642 or resi 643 or resi 644 or resi 646 or resi 646 or resi 648 or resi 649 or resi 649 or resi 652 or resi 653 or resi 653 or resi 654 or resi 662 or resi 664 or resi 669 or resi 684 or resi 685 or resi 690 or resi 694 or resi 695 or resi 705 or resi 712 or resi 713 or resi 714 or resi 716 or resi 716 or resi 727 or resi 731 or resi 731 or resi 733 or resi 734 or resi 740 or resi 749 or resi 753 or resi 760 or resi 760 or resi 762 or resi 772 or resi 776 or resi 780 or resi 803 or resi 804 or resi 809 or resi 812 or resi 814 or resi 817 or resi 817 or resi 817 or resi 818 or resi 820 or resi 888 or resi 889 or resi 920 or resi 933 or resi 935 or resi 962 or resi 968 or resi 989 or resi 1058 or resi 1073 or resi 1074 or resi 1132 or resi 1133 or resi 1150 or resi 1153 or resi 1175 or resi 1223 or resi 1251 or resi 1366 or resi 1421 or resi 1426 or resi 1448 or resi 1459 or resi 1462'"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#GRIN 2A LLISTA DISEASE ASSOCIATED\n",
    "#creo una llista amb la posició del aminoàcid\n",
    "llista1=list(G2A_da['Amino acid position'])\n",
    "#la ordeno per ordre creixent\n",
    "llista1.sort()\n",
    "#creo llista buida\n",
    "L1=[]\n",
    "#inicialitzo comptador a 0\n",
    "comp=0\n",
    "#mentres que la llargada de la llista sigui diferent al comptador\n",
    "while len(llista1)!=comp:\n",
    "    #poso que \"h\" és resi, la posició de la llista de aa, i or\n",
    "    h=['resi',llista1[comp],'or ']\n",
    "    #afegeixo h a la llista buida L1\n",
    "    L1.append(h)\n",
    "    #sumo 1 al comptador per canviar de posició\n",
    "    comp=comp+1\n",
    "#paso la llista a cadena\n",
    "s = ','.join(map(str, L1))\n",
    "#elimino: [ ] , '\n",
    "ss=s.replace('[','')\n",
    "sss=ss.replace(',','')\n",
    "ssss=sss.replace(']','')\n",
    "sssss=ssss.replace(\"'\",\"\")\n",
    "#ara borrem el \"or\" final \n",
    "Grin2A_da_aa = remove_last_or(sssss)\n",
    "Grin2A_da_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'resi 140 or resi 143 or resi 149 or resi 159 or resi 183 or resi 189 or resi 189 or resi 208 or resi 290 or resi 336 or resi 370 or resi 390 or resi 439 or resi 441 or resi 447 or resi 452 or resi 461 or resi 504 or resi 506 or resi 506 or resi 545 or resi 554 or resi 556 or resi 568 or resi 568 or resi 590 or resi 595 or resi 599 or resi 616 or resi 638 or resi 649 or resi 674 or resi 699 or resi 701 or resi 702 or resi 728 or resi 731 or resi 763 or resi 813 or resi 818 or resi 823 or resi 840 or resi 852 or resi 852 or resi 876 or resi 884 or resi 904 or resi 922 or resi 929 or resi 937 or resi 964 or resi 976 or resi 998 or resi 1000 or resi 1064 or resi 1076 or resi 1129 or resi 1193 or resi 1229 or resi 1276 or resi 1285 or resi 1318 or resi 1321 or resi 1402 or resi 1425'"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#GRIN 2A LLISTA Uncertain pathogenesis\n",
    "#creo una llista amb la posició del aminoàcid\n",
    "llista1=list(G2A_up['Amino acid position'])\n",
    "#la ordeno per ordre creixent\n",
    "llista1.sort()\n",
    "#creo llista buida\n",
    "L1=[]\n",
    "#inicialitzo comptador a 0\n",
    "comp=0\n",
    "#mentres que la llargada de la llista sigui diferent al comptador\n",
    "while len(llista1)!=comp:\n",
    "    #poso que \"h\" és resi, la posició de la llista de aa, i or\n",
    "    h=['resi',llista1[comp],'or ']\n",
    "    #afegeixo h a la llista buida L1\n",
    "    L1.append(h)\n",
    "    #sumo 1 al comptador per canviar de posició\n",
    "    comp=comp+1\n",
    "#paso la llista a cadena\n",
    "s = ','.join(map(str, L1))\n",
    "#elimino: [ ] , '\n",
    "ss=s.replace('[','')\n",
    "sss=ss.replace(',','')\n",
    "ssss=sss.replace(']','')\n",
    "sssss=ssss.replace(\"'\",\"\")\n",
    "#ara borrem el \"or\" final \n",
    "Grin2A_up_aa = remove_last_or(sssss)\n",
    "Grin2A_up_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'resi 5 or resi 8 or resi 11 or resi 13 or resi 19 or resi 20 or resi 20 or resi 21 or resi 23 or resi 24 or resi 25 or resi 26 or resi 26 or resi 27 or resi 28 or resi 29 or resi 31 or resi 32 or resi 33 or resi 40 or resi 45 or resi 45 or resi 46 or resi 47 or resi 49 or resi 49 or resi 51 or resi 54 or resi 56 or resi 58 or resi 59 or resi 60 or resi 60 or resi 61 or resi 61 or resi 64 or resi 65 or resi 68 or resi 69 or resi 70 or resi 73 or resi 74 or resi 74 or resi 76 or resi 76 or resi 80 or resi 80 or resi 80 or resi 83 or resi 84 or resi 88 or resi 90 or resi 90 or resi 92 or resi 93 or resi 93 or resi 94 or resi 94 or resi 95 or resi 96 or resi 99 or resi 99 or resi 102 or resi 112 or resi 116 or resi 122 or resi 124 or resi 133 or resi 134 or resi 136 or resi 138 or resi 140 or resi 141 or resi 141 or resi 147 or resi 153 or resi 155 or resi 156 or resi 161 or resi 166 or resi 166 or resi 187 or resi 190 or resi 192 or resi 193 or resi 195 or resi 199 or resi 200 or resi 200 or resi 201 or resi 201 or resi 203 or resi 203 or resi 204 or resi 206 or resi 209 or resi 211 or resi 212 or resi 213 or resi 221 or resi 223 or resi 223 or resi 228 or resi 230 or resi 232 or resi 234 or resi 238 or resi 244 or resi 244 or resi 245 or resi 246 or resi 247 or resi 249 or resi 250 or resi 253 or resi 265 or resi 267 or resi 272 or resi 274 or resi 276 or resi 277 or resi 279 or resi 289 or resi 297 or resi 297 or resi 298 or resi 300 or resi 302 or resi 304 or resi 305 or resi 306 or resi 306 or resi 308 or resi 309 or resi 311 or resi 313 or resi 315 or resi 316 or resi 317 or resi 319 or resi 319 or resi 321 or resi 322 or resi 324 or resi 327 or resi 327 or resi 328 or resi 329 or resi 330 or resi 330 or resi 333 or resi 333 or resi 335 or resi 336 or resi 336 or resi 339 or resi 340 or resi 340 or resi 343 or resi 343 or resi 344 or resi 347 or resi 353 or resi 354 or resi 356 or resi 357 or resi 358 or resi 358 or resi 362 or resi 365 or resi 367 or resi 369 or resi 375 or resi 381 or resi 382 or resi 382 or resi 387 or resi 388 or resi 389 or resi 400 or resi 401 or resi 414 or resi 417 or resi 422 or resi 423 or resi 424 or resi 424 or resi 424 or resi 425 or resi 427 or resi 431 or resi 432 or resi 437 or resi 438 or resi 439 or resi 440 or resi 440 or resi 440 or resi 441 or resi 444 or resi 447 or resi 448 or resi 448 or resi 449 or resi 449 or resi 451 or resi 457 or resi 459 or resi 462 or resi 463 or resi 470 or resi 471 or resi 476 or resi 478 or resi 481 or resi 482 or resi 491 or resi 498 or resi 506 or resi 520 or resi 521 or resi 546 or resi 556 or resi 559 or resi 564 or resi 571 or resi 576 or resi 580 or resi 583 or resi 586 or resi 589 or resi 589 or resi 591 or resi 594 or resi 594 or resi 596 or resi 601 or resi 607 or resi 630 or resi 671 or resi 673 or resi 675 or resi 675 or resi 676 or resi 678 or resi 681 or resi 682 or resi 701 or resi 706 or resi 719 or resi 722 or resi 722 or resi 734 or resi 742 or resi 743 or resi 744 or resi 750 or resi 755 or resi 756 or resi 768 or resi 773 or resi 775 or resi 776 or resi 783 or resi 784 or resi 788 or resi 811 or resi 818 or resi 820 or resi 827 or resi 828 or resi 832 or resi 832 or resi 833 or resi 836 or resi 836 or resi 839 or resi 839 or resi 846 or resi 848 or resi 849 or resi 850 or resi 851 or resi 853 or resi 855 or resi 855 or resi 856 or resi 856 or resi 856 or resi 857 or resi 858 or resi 861 or resi 865 or resi 867 or resi 867 or resi 869 or resi 876 or resi 876 or resi 876 or resi 876 or resi 879 or resi 880 or resi 883 or resi 884 or resi 885 or resi 888 or resi 889 or resi 892 or resi 894 or resi 894 or resi 894 or resi 894 or resi 896 or resi 899 or resi 899 or resi 901 or resi 904 or resi 905 or resi 906 or resi 909 or resi 910 or resi 910 or resi 912 or resi 915 or resi 916 or resi 916 or resi 917 or resi 918 or resi 922 or resi 924 or resi 927 or resi 927 or resi 931 or resi 931 or resi 932 or resi 932 or resi 937 or resi 940 or resi 941 or resi 942 or resi 943 or resi 945 or resi 945 or resi 951 or resi 953 or resi 954 or resi 955 or resi 959 or resi 960 or resi 964 or resi 965 or resi 967 or resi 969 or resi 970 or resi 970 or resi 973 or resi 977 or resi 982 or resi 987 or resi 995 or resi 995 or resi 997 or resi 999 or resi 1002 or resi 1002 or resi 1002 or resi 1005 or resi 1006 or resi 1010 or resi 1011 or resi 1011 or resi 1014 or resi 1018 or resi 1021 or resi 1022 or resi 1022 or resi 1023 or resi 1027 or resi 1028 or resi 1029 or resi 1031 or resi 1035 or resi 1040 or resi 1040 or resi 1041 or resi 1041 or resi 1042 or resi 1046 or resi 1047 or resi 1051 or resi 1055 or resi 1055 or resi 1056 or resi 1057 or resi 1059 or resi 1061 or resi 1064 or resi 1064 or resi 1065 or resi 1069 or resi 1072 or resi 1075 or resi 1077 or resi 1077 or resi 1077 or resi 1081 or resi 1082 or resi 1083 or resi 1084 or resi 1088 or resi 1089 or resi 1090 or resi 1091 or resi 1098 or resi 1100 or resi 1100 or resi 1102 or resi 1103 or resi 1103 or resi 1104 or resi 1106 or resi 1111 or resi 1111 or resi 1111 or resi 1113 or resi 1115 or resi 1115 or resi 1115 or resi 1117 or resi 1121 or resi 1121 or resi 1121 or resi 1122 or resi 1123 or resi 1123 or resi 1125 or resi 1126 or resi 1126 or resi 1131 or resi 1132 or resi 1133 or resi 1134 or resi 1137 or resi 1137 or resi 1139 or resi 1140 or resi 1143 or resi 1144 or resi 1145 or resi 1148 or resi 1148 or resi 1151 or resi 1152 or resi 1155 or resi 1156 or resi 1156 or resi 1157 or resi 1159 or resi 1159 or resi 1159 or resi 1160 or resi 1162 or resi 1168 or resi 1169 or resi 1169 or resi 1170 or resi 1171 or resi 1172 or resi 1173 or resi 1174 or resi 1177 or resi 1178 or resi 1179 or resi 1179 or resi 1180 or resi 1182 or resi 1182 or resi 1186 or resi 1190 or resi 1191 or resi 1192 or resi 1193 or resi 1194 or resi 1200 or resi 1201 or resi 1201 or resi 1203 or resi 1206 or resi 1207 or resi 1208 or resi 1212 or resi 1213 or resi 1214 or resi 1216 or resi 1216 or resi 1221 or resi 1222 or resi 1222 or resi 1223 or resi 1223 or resi 1227 or resi 1228 or resi 1229 or resi 1234 or resi 1237 or resi 1241 or resi 1241 or resi 1242 or resi 1243 or resi 1243 or resi 1246 or resi 1249 or resi 1249 or resi 1252 or resi 1256 or resi 1262 or resi 1263 or resi 1265 or resi 1265 or resi 1266 or resi 1266 or resi 1266 or resi 1268 or resi 1268 or resi 1270 or resi 1270 or resi 1272 or resi 1273 or resi 1273 or resi 1274 or resi 1275 or resi 1276 or resi 1277 or resi 1277 or resi 1278 or resi 1281 or resi 1281 or resi 1282 or resi 1284 or resi 1285 or resi 1288 or resi 1290 or resi 1291 or resi 1292 or resi 1293 or resi 1293 or resi 1293 or resi 1294 or resi 1295 or resi 1297 or resi 1297 or resi 1300 or resi 1300 or resi 1303 or resi 1304 or resi 1305 or resi 1306 or resi 1306 or resi 1307 or resi 1309 or resi 1310 or resi 1311 or resi 1312 or resi 1313 or resi 1315 or resi 1315 or resi 1318 or resi 1318 or resi 1326 or resi 1329 or resi 1334 or resi 1338 or resi 1339 or resi 1340 or resi 1340 or resi 1341 or resi 1341 or resi 1343 or resi 1352 or resi 1353 or resi 1357 or resi 1359 or resi 1361 or resi 1362 or resi 1366 or resi 1370 or resi 1370 or resi 1370 or resi 1372 or resi 1376 or resi 1376 or resi 1377 or resi 1377 or resi 1382 or resi 1382 or resi 1384 or resi 1385 or resi 1388 or resi 1390 or resi 1390 or resi 1392 or resi 1395 or resi 1396 or resi 1402 or resi 1403 or resi 1405 or resi 1407 or resi 1408 or resi 1408 or resi 1412 or resi 1414 or resi 1415 or resi 1416 or resi 1417 or resi 1417 or resi 1420 or resi 1420 or resi 1420 or resi 1426 or resi 1436 or resi 1437 or resi 1438 or resi 1438 or resi 1440 or resi 1441 or resi 1442 or resi 1444 or resi 1445 or resi 1447 or resi 1448 or resi 1448 or resi 1451 or resi 1452 or resi 1452 or resi 1452 or resi 1453 or resi 1460 or resi 1460 or resi 1461 or resi 1463'"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#GRIN 2A LLISTA Neutral\n",
    "#creo una llista amb la posició del aminoàcid\n",
    "llista1=list(G2A_n['Amino acid position'])\n",
    "#la ordeno per ordre creixent\n",
    "llista1.sort()\n",
    "#creo llista buida\n",
    "L1=[]\n",
    "#inicialitzo comptador a 0\n",
    "comp=0\n",
    "#mentres que la llargada de la llista sigui diferent al comptador\n",
    "while len(llista1)!=comp:\n",
    "    #poso que \"h\" és resi, la posició de la llista de aa, i or\n",
    "    h=['resi',llista1[comp],'or ']\n",
    "    #afegeixo h a la llista buida L1\n",
    "    L1.append(h)\n",
    "    #sumo 1 al comptador per canviar de posició\n",
    "    comp=comp+1\n",
    "#paso la llista a cadena\n",
    "s = ','.join(map(str, L1))\n",
    "#elimino: [ ] , '\n",
    "ss=s.replace('[','')\n",
    "sss=ss.replace(',','')\n",
    "ssss=sss.replace(']','')\n",
    "sssss=ssss.replace(\"'\",\"\")\n",
    "#ara borrem el \"or\" final \n",
    "Grin2A_n_aa = remove_last_or(sssss)\n",
    "Grin2A_n_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "llista dels aa que estan a gnomAD V2 i no a GRIN db:  [['Ser', '1425', 'Leu'], ['Ala', '1395', 'Glu'], ['Asp', '1364', 'Asn'], ['Glu', '1321', 'Gln'], ['Arg', '1318', 'Trp'], ['Asp', '1297', 'Asn'], ['Arg', '1285', 'Lys'], ['Ala', '1276', 'Gly'], ['Met', '1242', 'Lys'], ['Thr', '1229', 'Ser'], ['Leu', '1193', 'Trp'], ['Pro', '1171', 'Ser'], ['Pro', '1142', 'Ser'], ['His', '1129', 'Arg'], ['Gly', '1127', 'Ser'], ['Arg', '1114', 'Ser'], ['Lys', '1107', 'Ile'], ['Thr', '1104', 'Asn'], ['Asn', '1076', 'Lys'], ['Asn', '1076', 'Lys'], ['Thr', '1064', 'Ala'], ['His', '1044', 'Arg'], ['Asn', '1029', 'Thr'], ['Lys', '1015', 'Arg'], ['Val', '998', 'Met'], ['Val', '996', 'Met'], ['Asn', '989', 'Ser'], ['Asn', '976', 'Ser'], ['Gln', '964', 'Glu'], ['Asp', '937', 'Asn'], ['Met', '934', 'Lys'], ['Ser', '929', 'Phe'], ['Ala', '922', 'Val'], ['Ala', '921', 'Thr'], ['Asp', '884', 'Asn'], ['Ile', '876', 'Thr'], ['Val', '852', 'Met'], ['Arg', '846', 'Cys'], ['Leu', '840', 'Phe'], ['Val', '820', 'Ile'], ['Ile', '814', 'Thr'], ['Glu', '803', 'Lys'], ['Ala', '716', 'Thr'], ['Thr', '690', 'Met'], ['His', '674', 'Gln'], ['Trp', '606', 'Cys'], ['Ile', '605', 'Met'], ['His', '595', 'Arg'], ['Lys', '590', 'Asn'], ['Val', '573', 'Leu'], ['Val', '568', 'Leu'], ['Pro', '527', 'Leu'], ['Arg', '518', 'Cys'], ['Ala', '508', 'Val'], ['Val', '506', 'Ala'], ['Arg', '504', 'Trp'], ['Val', '452', 'Met'], ['Asn', '447', 'Lys'], ['Pro', '435', 'Leu'], ['Glu', '427', 'Lys'], ['Arg', '370', 'Trp'], ['Val', '357', 'Leu'], ['Pro', '336', 'Ser'], ['Gly', '295', 'Ser'], ['Ala', '290', 'Val'], ['Leu', '248', 'Val'], ['Thr', '208', 'Ile'], ['Thr', '189', 'Asn'], ['Thr', '189', 'Ala'], ['Phe', '183', 'Ile'], ['Pro', '178', 'Thr'], ['His', '168', 'Asn'], ['His', '168', 'Tyr'], ['Ala', '149', 'Val'], ['Phe', '144', 'Ser'], ['Pro', '140', 'Ala'], ['Asp', '139', 'Gly'], ['Arg', '94', 'Leu'], ['Val', '86', 'Met'], ['Pro', '79', 'Arg'], ['His', '42', 'Gln'], ['Pro', '31', 'Thr'], ['Ala', '27', 'Thr'], ['Ala', '26', 'Thr'], ['Pro', '21', 'Leu'], ['Arg', '19', 'His'], ['Pro', '13', 'Thr'], ['Leu', '12', 'Met'], ['Val', '11', 'Met'], ['Leu', '9', 'Met'], ['Gly', '5', 'Val'], ['Gly', '2', 'Ser']]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ys/tvdvp6kj24df4dh9qb_87_tr0000gn/T/ipykernel_43013/1568932411.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2A=G2A_rev[['lletres', 'numeros']] = G2A_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n",
      "/var/folders/ys/tvdvp6kj24df4dh9qb_87_tr0000gn/T/ipykernel_43013/1568932411.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2A=G2A_rev[['lletres', 'numeros']] = G2A_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n"
     ]
    }
   ],
   "source": [
    "#COMPARACIÓ GRIN2A GRINDB VS GNOMAD V2\n",
    "\n",
    "#De les que tinc a la base de dades\n",
    "#Per fer Met654Ile de les neutral\n",
    "#copio la llista de grin1 neutral\n",
    "G2A_aa=G2A_n\n",
    "#alimino totes les columnes que no necessito\n",
    "G2A_aa=G2A_aa.drop(['Gene','Gain/Loss','Disease-association','Mutation Category'], axis=1)\n",
    "#les reordeno\n",
    "G2A_aa = G2A_aa.reindex(columns=['Reference Amino acid','Amino acid position','Mutated Amino acid'])\n",
    "#Poso com string la posició del aa mutat\n",
    "G2A_aa['string_numbers'] = G2A_aa['Amino acid position'].astype(str)\n",
    "#alimino totes les columnes que no necessito\n",
    "G2A_aa=G2A_aa.drop(['Amino acid position'], axis=1)\n",
    "#les reordeno\n",
    "G2A_aa = G2A_aa.reindex(columns=['Reference Amino acid','string_numbers','Mutated Amino acid'])\n",
    "#ho paso a llista\n",
    "G2A_aaL=G2A_aa.to_numpy().tolist()\n",
    "#ara tinc la llista dels aa mutació neutra de grin1 que estan a la base de dades\n",
    "G2A_aaL\n",
    "\n",
    "\n",
    "#llegeixo arxiu gnomad v2\n",
    "dades = pd.read_csv(\"/Users/ismaelvera/Desktop/TFG/Mutacions GRINdb/GRIN_missense_variants_gnomad_v2_1.txt\",header=None) \n",
    "#em quedo nomes amb les columnes que diuen la mutació i el gen\n",
    "dades=dades.drop([0,1,2,3,4,5,6,8,10,11,12,13,14], axis=1)\n",
    "#faig una llista amb els que son nomes de grin1\n",
    "G2A_rev=dades.loc[dades[9]=='GRIN2A']\n",
    "#com que ho tinc com a p.Ser2Arg\n",
    "#creo funció per separar lletres de numeros \n",
    "def separate_letters_numbers(text):\n",
    "    lletres = ''.join(filter(str.isalpha, text))\n",
    "    numeros = ''.join(filter(str.isdigit, text))\n",
    "    return (lletres, numeros)\n",
    "#aplico funció a la columna que m'interesa\n",
    "df2A=G2A_rev[['lletres', 'numeros']] = G2A_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n",
    "#ara tinc lletres i numeros separats\n",
    "#separo les lletres per tenir p, Ser, Arg\n",
    "df2A['col1'] = df2A[0].str[0]\n",
    "df2A['col2'] = df2A[0].str[1:4]\n",
    "df2A['col3'] = df2A[0].str[4:]\n",
    "#elimino totes les columnes que no necessito\n",
    "df2A=df2A.drop([0,'col1'], axis=1)\n",
    "#les reordeno\n",
    "df2A = df2A.reindex(columns=['col2',1,'col3'])\n",
    "#ho paso a llista\n",
    "df2L2A=df2A.to_numpy().tolist()\n",
    "\n",
    "\n",
    "#faig la comparació\n",
    "#he de crear una llista buida on guardare els que no estan \n",
    "noestanG2AV2=[]\n",
    "#per cada element que tinc a la llista de gnomad v2, si no esta a la llista dels \n",
    "# que tinc a la base de dades, mel afegeix a la llista L1\n",
    "for element in df2L2A:\n",
    "    if element not in G2A_aaL:\n",
    "        noestanG2AV2.append(element)\n",
    "\n",
    "\n",
    "#ara per poder marcar-los al model molecular he de fer que posi resi 23 or resi 234 ....\n",
    "#primer separo la llista L1, que son els que no tinc en AA1 posició AA2\n",
    "df2A_2 = pd.DataFrame(noestanG2AV2, columns=['AA 1', 'Position', 'AA 2'])\n",
    "#creo una llista amb la posició del aminoàcid\n",
    "llista1=list(df2A_2['Position'])\n",
    "#la ordeno per ordre creixent\n",
    "llista1.sort()\n",
    "#creo llista buida\n",
    "L1=[]\n",
    "#inicialitzo comptador a 0\n",
    "comp=0\n",
    "#mentres que la llargada de la llista sigui diferent al comptador\n",
    "while len(llista1)!=comp:\n",
    "    #poso que \"h\" és resi, la posició de la llista de aa, i or\n",
    "    h=['resi',llista1[comp],'or ']\n",
    "    #afegeixo h a la llista buida L1\n",
    "    L1.append(h)\n",
    "    #sumo 1 al comptador per canviar de posició\n",
    "    comp=comp+1\n",
    "#paso la llista a cadena\n",
    "s = ','.join(map(str, L1))\n",
    "#elimino: [ ] , '\n",
    "ss=s.replace('[','')\n",
    "sss=ss.replace(',','')\n",
    "ssss=sss.replace(']','')\n",
    "sssss=ssss.replace(\"'\",\"\")\n",
    "#ara borrem el \"or\" final \n",
    "#faig una funció per borrar el or final\n",
    "def remove_last_or(chain):\n",
    "    #Divideix la cadena en diferents elements separats per \"or\"\n",
    "    elements = chain.split(\"or\")\n",
    "    #Per si nomes hagues un a la llista, retorna la cadena tal com està\n",
    "    if len(elements) == 1:\n",
    "        return chain\n",
    "    #Treu qualsevol espai en blanc inicial o final de cada element\n",
    "    cleaned_elements = [elem.strip() for elem in elements]\n",
    "    #Elimina l'últim element, que hauria de ser una cadena buida després de l'últim \"or\"\n",
    "    cleaned_elements.pop()\n",
    "    #Torna a unir els elements \"netejats\" fent servir \"or\" com a separador\n",
    "    new_chain = \" or \".join(cleaned_elements)\n",
    "    return new_chain\n",
    "G2A_rev_v2 = remove_last_or(sssss)\n",
    "G2A_rev_v2\n",
    "\n",
    "#mostro la llista dels que no estan\n",
    "print('llista dels aa que estan a gnomAD V2 i no a GRIN db: ',noestanG2AV2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "llista dels aa que estan a gnomAD V2 i no a GRIN db:  [['Arg', '1451', 'Ile'], ['Ser', '1440', 'Pro'], ['Val', '1428', 'Ala'], ['Asn', '1420', 'His'], ['His', '1419', 'Gln'], ['Ser', '1416', 'Ile'], ['Arg', '1406', 'Lys'], ['Arg', '1402', 'Pro'], ['Arg', '1124', 'Gln'], ['Gln', '1394', 'Pro'], ['His', '1121', 'Arg'], ['Arg', '1119', 'His'], ['Ser', '1390', 'Pro'], ['Gly', '1109', 'Glu'], ['Arg', '1376', 'Ser'], ['Pro', '1359', 'Leu'], ['Lys', '1355', 'Asn'], ['Gly', '1347', 'Ser'], ['Ser', '1337', 'Leu'], ['Lys', '1335', 'Glu'], ['Glu', '1321', 'Gln'], ['Arg', '1318', 'Trp'], ['Ser', '1305', 'Thr'], ['Glu', '1301', 'Asp'], ['Ile', '1295', 'Asn'], ['Ile', '1295', 'Val'], ['Ser', '1291', 'Cys'], ['Arg', '1285', 'Lys'], ['Ala', '1276', 'Gly'], ['Gln', '1273', 'Arg'], ['Gln', '1265', 'Glu'], ['Glu', '1264', 'Lys'], ['Thr', '1262', 'Ile'], ['Met', '1242', 'Ile'], ['Leu', '1240', 'Gln'], ['Pro', '1233', 'Ser'], ['Met', '1230', 'Val'], ['Thr', '1229', 'Ser'], ['Met', '1221', 'Val'], ['His', '1213', 'Arg'], ['Thr', '1212', 'Met'], ['Arg', '1208', 'Trp'], ['Ser', '1204', 'Arg'], ['Ser', '1204', 'Asn'], ['Pro', '1199', 'Ser'], ['Leu', '1193', 'Trp'], ['His', '1190', 'Pro'], ['His', '1190', 'Leu'], ['Asn', '1180', 'Ile'], ['Asn', '1180', 'Ser'], ['His', '1173', 'Gln'], ['Lys', '1160', 'Asn'], ['Pro', '1154', 'Leu'], ['Gln', '1152', 'His'], ['Pro', '1142', 'Ser'], ['Asp', '1131', 'Glu'], ['His', '1129', 'Arg'], ['His', '1129', 'Leu'], ['Phe', '1128', 'Leu'], ['Lys', '1124', 'Thr'], ['Asp', '1121', 'Ala'], ['Asp', '1121', 'His'], ['Lys', '1116', 'Glu'], ['Asp', '1115', 'Ala'], ['Asp', '1115', 'Asn'], ['Arg', '1114', 'Ser'], ['Arg', '1114', 'Thr'], ['Pro', '1113', 'Ala'], ['Ser', '1112', 'Tyr'], ['Ser', '1112', 'Ala'], ['Ser', '1112', 'Thr'], ['Ser', '1111', 'Asn'], ['Ser', '1111', 'Gly'], ['Ser', '1111', 'Cys'], ['Ser', '1110', 'Leu'], ['Ser', '1110', 'Thr'], ['Thr', '1108', 'Ile'], ['Thr', '1108', 'Asn'], ['Thr', '1108', 'Ser'], ['Lys', '1107', 'Ile'], ['Lys', '1107', 'Arg'], ['Lys', '1107', 'Glu'], ['Leu', '1106', 'Met'], ['Tyr', '1105', 'Phe'], ['Tyr', '1105', 'Ser'], ['Tyr', '1105', 'Asn'], ['Tyr', '1105', 'His'], ['Tyr', '1105', 'Asp'], ['Thr', '1104', 'Asn'], ['Thr', '1104', 'Pro'], ['Arg', '1103', 'Leu'], ['Ser', '1099', 'Arg'], ['Lys', '1096', 'Asn'], ['Ala', '1091', 'Thr'], ['Val', '1090', 'Leu'], ['Lys', '1087', 'Asn'], ['Lys', '1078', 'Glu'], ['Asn', '1076', 'Lys'], ['His', '1071', 'Tyr'], ['Arg', '1067', 'Gln'], ['Thr', '1064', 'Ala'], ['His', '1058', 'Leu'], ['Met', '1056', 'Val'], ['Ser', '1048', 'Arg'], ['Lys', '1047', 'Glu'], ['His', '1044', 'Arg'], ['Asn', '1041', 'Ser'], ['Thr', '1038', 'Ile'], ['Glu', '1036', 'Val'], ['Leu', '1026', 'Val'], ['Asp', '1019', 'Asn'], ['Pro', '1010', 'Ser'], ['Ser', '1008', 'Cys'], ['Lys', '1005', 'Thr'], ['Val', '998', 'Ala'], ['Asn', '992', 'Lys'], ['Asn', '989', 'Ser'], ['Val', '979', 'Ile'], ['Asn', '976', 'Lys'], ['Arg', '970', 'Leu'], ['Leu', '963', 'Phe'], ['Ile', '955', 'Met'], ['Ile', '955', 'Asn'], ['Ser', '954', 'Arg'], ['Asp', '937', 'Asn'], ['Val', '935', 'Phe'], ['Met', '934', 'Lys'], ['Met', '932', 'Thr'], ['Ile', '931', 'Leu'], ['Ala', '922', 'Val'], ['Ala', '922', 'Thr'], ['Ser', '890', 'Pro'], ['Thr', '888', 'Arg'], ['Asp', '884', 'Asn'], ['Ser', '882', 'Pro'], ['Ser', '882', 'Thr'], ['Ile', '876', 'Thr'], ['His', '796', 'Tyr'], ['Trp', '793', 'Arg'], ['Cys', '870', 'Tyr'], ['Val', '852', 'Met'], ['Gly', '851', 'Ala'], ['Leu', '840', 'Phe'], ['Ala', '825', 'Asp'], ['Gly', '819', 'Ser'], ['Ile', '814', 'Thr'], ['Asn', '802', 'Ser'], ['Gln', '781', 'His'], ['Ile', '763', 'Met'], ['Asp', '726', 'Glu'], ['Gly', '723', 'Arg'], ['His', '674', 'Gln'], ['Met', '630', 'Ile'], ['Ile', '605', 'Met'], ['Lys', '590', 'Asn'], ['Ile', '571', 'Val'], ['Ser', '534', 'Asn'], ['Pro', '527', 'Ser'], ['Met', '507', 'Thr'], ['Val', '506', 'Ala'], ['Val', '506', 'Ile'], ['Arg', '504', 'Gln'], ['Gln', '503', 'Lys'], ['Tyr', '502', 'Phe'], ['Val', '501', 'Gly'], ['Val', '500', 'Gly'], ['Lys', '472', 'Gln'], ['Thr', '470', 'Ile'], ['Val', '452', 'Met'], ['Met', '450', 'Leu'], ['Glu', '448', 'Gly'], ['Asn', '447', 'Lys'], ['Asn', '447', 'Ser'], ['Arg', '437', 'Trp'], ['Ile', '408', 'Thr'], ['Glu', '400', 'Val'], ['Lys', '394', 'Met'], ['Lys', '394', 'Gln'], ['Ser', '384', 'Arg'], ['Ser', '384', 'Thr'], ['His', '358', 'Asp'], ['Tyr', '355', 'Phe'], ['Asp', '347', 'Val'], ['Asp', '344', 'Asn'], ['Val', '339', 'Gly'], ['Pro', '336', 'Ser'], ['Pro', '330', 'Ser'], ['Ile', '313', 'Leu'], ['Ile', '313', 'Val'], ['Phe', '310', 'Leu'], ['Glu', '308', 'Lys'], ['Gly', '295', 'Ser'], ['Ala', '290', 'Val'], ['Ser', '287', 'Arg'], ['Ser', '280', 'Phe'], ['Glu', '271', 'Gln'], ['Pro', '258', 'Ser'], ['Tyr', '251', 'Phe'], ['Thr', '249', 'Asn'], ['Glu', '242', 'Asp'], ['Leu', '229', 'Phe'], ['His', '223', 'Arg'], ['Thr', '208', 'Ile'], ['Gly', '197', 'Cys'], ['Phe', '186', 'Ser'], ['Phe', '183', 'Ile'], ['Ile', '176', 'Val'], ['Val', '173', 'Leu'], ['Thr', '156', 'Ala'], ['Gly', '130', 'Asp'], ['Gln', '106', 'His'], ['His', '96', 'Tyr'], ['Arg', '94', 'Cys'], ['Ala', '71', 'Gly'], ['Val', '67', 'Met'], ['Glu', '58', 'Asp'], ['Arg', '52', 'Leu'], ['Arg', '52', 'Gln'], ['Ala', '27', 'Pro'], ['Ala', '25', 'Thr'], ['Ala', '25', 'Ser'], ['Ser', '24', 'Asn'], ['Pro', '23', 'Leu'], ['Ala', '22', 'Pro'], ['Pro', '21', 'Leu'], ['Val', '17', 'Ala']]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ys/tvdvp6kj24df4dh9qb_87_tr0000gn/T/ipykernel_43013/4229322101.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2A=G2A_rev[['lletres', 'numeros']] = G2A_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n",
      "/var/folders/ys/tvdvp6kj24df4dh9qb_87_tr0000gn/T/ipykernel_43013/4229322101.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2A=G2A_rev[['lletres', 'numeros']] = G2A_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n"
     ]
    }
   ],
   "source": [
    "#COMPARACIÓ GRIN2A GRINDB VS GNOMAD V3\n",
    "\n",
    "#De les que tinc a la base de dades\n",
    "#Per fer Met654Ile de les neutral\n",
    "#copio la llista de grin2A neutral\n",
    "G2A_aa=G2A_n\n",
    "#alimino totes les columnes que no necessito\n",
    "G2A_aa=G2A_aa.drop(['Gene','Gain/Loss','Disease-association','Mutation Category'], axis=1)\n",
    "#les reordeno\n",
    "G2A_aa = G2A_aa.reindex(columns=['Reference Amino acid','Amino acid position','Mutated Amino acid'])\n",
    "#Poso com string la posició del aa mutat\n",
    "G2A_aa['string_numbers'] = G2A_aa['Amino acid position'].astype(str)\n",
    "#alimino totes les columnes que no necessito\n",
    "G2A_aa=G2A_aa.drop(['Amino acid position'], axis=1)\n",
    "#les reordeno\n",
    "G2A_aa = G2A_aa.reindex(columns=['Reference Amino acid','string_numbers','Mutated Amino acid'])\n",
    "#ho paso a llista\n",
    "G2A_aaL=G2A_aa.to_numpy().tolist()\n",
    "#ara tinc la llista dels aa mutació neutra de grin1 que estan a la base de dades\n",
    "G2A_aaL\n",
    "\n",
    "\n",
    "#llegeixo arxiu gnomad v2\n",
    "dades = pd.read_csv(\"/Users/ismaelvera/Desktop/TFG/Mutacions GRINdb/GRIN_missense_variants_gnomad_v3.txt\",header=None) \n",
    "#em quedo nomes amb les columnes que diuen la mutació i el gen\n",
    "dades=dades.drop([0,1,2,3,4,5,6,8,10,11,12,13,14], axis=1)\n",
    "#faig una llista amb els que son nomes de grin1\n",
    "G2A_rev=dades.loc[dades[9]=='GRIN2A']\n",
    "#com que ho tinc com a p.Ser2Arg\n",
    "#creo funció per separar lletres de numeros \n",
    "def separate_letters_numbers(text):\n",
    "    lletres = ''.join(filter(str.isalpha, text))\n",
    "    numeros = ''.join(filter(str.isdigit, text))\n",
    "    return (lletres, numeros)\n",
    "#aplico funció a la columna que m'interesa\n",
    "df2A=G2A_rev[['lletres', 'numeros']] = G2A_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n",
    "#ara tinc lletres i numeros separats\n",
    "#separo les lletres per tenir p, Ser, Arg\n",
    "df2A['col1'] = df2A[0].str[0]\n",
    "df2A['col2'] = df2A[0].str[1:4]\n",
    "df2A['col3'] = df2A[0].str[4:]\n",
    "#elimino totes les columnes que no necessito\n",
    "df2A=df2A.drop([0,'col1'], axis=1)\n",
    "#les reordeno\n",
    "df2A = df2A.reindex(columns=['col2',1,'col3'])\n",
    "#ho paso a llista\n",
    "df2L2A=df2A.to_numpy().tolist()\n",
    "\n",
    "\n",
    "#faig la comparació\n",
    "#he de crear una llista buida on guardare els que no estan \n",
    "noestanG2AV3=[]\n",
    "#per cada element que tinc a la llista de gnomad v2, si no esta a la llista dels \n",
    "# que tinc a la base de dades, mel afegeix a la llista L1\n",
    "for element in df2L2A:\n",
    "    if element not in G2A_aaL:\n",
    "        noestanG2AV3.append(element)\n",
    "\n",
    "\n",
    "#ara per poder marcar-los al model molecular he de fer que posi resi 23 or resi 234 ....\n",
    "#primer separo la llista L1, que son els que no tinc en AA1 posició AA2\n",
    "df2A_2 = pd.DataFrame(noestanG2AV3, columns=['AA 1', 'Position', 'AA 2'])\n",
    "#creo una llista amb la posició del aminoàcid\n",
    "llista1=list(df2A_2['Position'])\n",
    "#la ordeno per ordre creixent\n",
    "llista1.sort()\n",
    "#creo llista buida\n",
    "L1=[]\n",
    "#inicialitzo comptador a 0\n",
    "comp=0\n",
    "#mentres que la llargada de la llista sigui diferent al comptador\n",
    "while len(llista1)!=comp:\n",
    "    #poso que \"h\" és resi, la posició de la llista de aa, i or\n",
    "    h=['resi',llista1[comp],'or ']\n",
    "    #afegeixo h a la llista buida L1\n",
    "    L1.append(h)\n",
    "    #sumo 1 al comptador per canviar de posició\n",
    "    comp=comp+1\n",
    "#paso la llista a cadena\n",
    "s = ','.join(map(str, L1))\n",
    "#elimino: [ ] , '\n",
    "ss=s.replace('[','')\n",
    "sss=ss.replace(',','')\n",
    "ssss=sss.replace(']','')\n",
    "sssss=ssss.replace(\"'\",\"\")\n",
    "#ara borrem el \"or\" final \n",
    "#faig una funció per borrar el or final\n",
    "def remove_last_or(chain):\n",
    "    #Divideix la cadena en diferents elements separats per \"or\"\n",
    "    elements = chain.split(\"or\")\n",
    "    #Per si nomes hagues un a la llista, retorna la cadena tal com està\n",
    "    if len(elements) == 1:\n",
    "        return chain\n",
    "    #Treu qualsevol espai en blanc inicial o final de cada element\n",
    "    cleaned_elements = [elem.strip() for elem in elements]\n",
    "    #Elimina l'últim element, que hauria de ser una cadena buida després de l'últim \"or\"\n",
    "    cleaned_elements.pop()\n",
    "    #Torna a unir els elements \"netejats\" fent servir \"or\" com a separador\n",
    "    new_chain = \" or \".join(cleaned_elements)\n",
    "    return new_chain\n",
    "G2A_rev_v3 = remove_last_or(sssss)\n",
    "G2A_rev_v3\n",
    "\n",
    "#mostro la llista dels que no estan\n",
    "print('llista dels aa que estan a gnomAD V2 i no a GRIN db: ',noestanG2AV3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Per GRIN2B\n",
    "#Agafem de la taula només GRIN1 i creem nova taula\n",
    "G2B=taula2.loc[taula2['Gene']=='GRIN2B']\n",
    "#De la nova taula agafem només Missense i creem una altra nova taula\n",
    "G2B_1=G2B.loc[G2B['Mutation Category']=='Missense']\n",
    "#De la nova taula agafem només les Disease associated, Uncertain pathogenesis, i neutral creem una altra noves taules\n",
    "#Taula Disease associated\n",
    "G2B_da=G2B_1.loc[G2B['Disease-association']=='Disease associated']\n",
    "#Taula Uncertain pathogenesis\n",
    "G2B_up=G2B_1.loc[G2B['Disease-association']=='Uncertain pathogenesis']\n",
    "#Taula Neutral\n",
    "G2B_n=G2B_1.loc[G2B['Disease-association']=='Neutral']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'resi 3 or resi 15 or resi 150 or resi 413 or resi 427 or resi 436 or resi 436 or resi 456 or resi 457 or resi 459 or resi 461 or resi 484 or resi 486 or resi 499 or resi 513 or resi 513 or resi 514 or resi 516 or resi 519 or resi 519 or resi 524 or resi 525 or resi 526 or resi 528 or resi 532 or resi 533 or resi 533 or resi 540 or resi 541 or resi 543 or resi 543 or resi 543 or resi 549 or resi 550 or resi 551 or resi 553 or resi 553 or resi 555 or resi 555 or resi 558 or resi 559 or resi 607 or resi 607 or resi 608 or resi 611 or resi 615 or resi 615 or resi 616 or resi 618 or resi 620 or resi 628 or resi 628 or resi 628 or resi 636 or resi 636 or resi 639 or resi 641 or resi 643 or resi 647 or resi 649 or resi 649 or resi 652 or resi 652 or resi 655 or resi 657 or resi 662 or resi 668 or resi 668 or resi 676 or resi 682 or resi 683 or resi 686 or resi 687 or resi 687 or resi 689 or resi 689 or resi 691 or resi 693 or resi 693 or resi 694 or resi 695 or resi 695 or resi 696 or resi 696 or resi 706 or resi 706 or resi 732 or resi 732 or resi 732 or resi 732 or resi 734 or resi 737 or resi 739 or resi 739 or resi 742 or resi 746 or resi 750 or resi 751 or resi 751 or resi 753 or resi 764 or resi 768 or resi 778 or resi 781 or resi 789 or resi 796 or resi 799 or resi 807 or resi 810 or resi 810 or resi 818 or resi 818 or resi 819 or resi 820 or resi 820 or resi 820 or resi 820 or resi 821 or resi 824 or resi 824 or resi 825 or resi 826 or resi nan or resi 832 or resi 839 or resi 917 or resi 1342'"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#GRIN 2B LLISTA DISEASE ASSOCIATED\n",
    "#creo una llista amb la posició del aminoàcid\n",
    "llista1=list(G2B_da['Amino acid position'])\n",
    "#la ordeno per ordre creixent\n",
    "llista1.sort()\n",
    "#creo llista buida\n",
    "L1=[]\n",
    "#inicialitzo comptador a 0\n",
    "comp=0\n",
    "#mentres que la llargada de la llista sigui diferent al comptador\n",
    "while len(llista1)!=comp:\n",
    "    #poso que \"h\" és resi, la posició de la llista de aa, i or\n",
    "    h=['resi',llista1[comp],'or ']\n",
    "    #afegeixo h a la llista buida L1\n",
    "    L1.append(h)\n",
    "    #sumo 1 al comptador per canviar de posició\n",
    "    comp=comp+1\n",
    "#paso la llista a cadena\n",
    "s = ','.join(map(str, L1))\n",
    "#elimino: [ ] , '\n",
    "ss=s.replace('[','')\n",
    "sss=ss.replace(',','')\n",
    "ssss=sss.replace(']','')\n",
    "sssss=ssss.replace(\"'\",\"\")\n",
    "#ara borrem el \"or\" final \n",
    "Grin2B_da_aa = remove_last_or(sssss)\n",
    "Grin2B_da_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'resi 8 or resi 18 or resi 42 or resi 50 or resi 83 or resi 126 or resi 156 or resi 157 or resi 181 or resi 181 or resi 271 or resi 293 or resi 362 or resi 407 or resi 514 or resi 533 or resi 541 or resi 554 or resi 559 or resi 617 or resi 646 or resi 656 or resi 660 or resi 682 or resi 684 or resi 684 or resi 685 or resi 690 or resi 691 or resi 696 or resi 733 or resi 734 or resi 817 or resi 818 or resi 826 or resi 829 or resi 837 or resi 839 or resi 844 or resi 848 or resi 888 or resi 952 or resi 993 or resi 1011 or resi 1014 or resi 1026 or resi 1081 or resi 1090 or resi 1111 or resi 1292 or resi 1415 or resi 1424 or resi 1439 or resi 1446 or resi 1452'"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#GRIN 2B LLISTA Uncertain pathogenesis\n",
    "#creo una llista amb la posició del aminoàcid\n",
    "llista1=list(G2B_up['Amino acid position'])\n",
    "#la ordeno per ordre creixent\n",
    "llista1.sort()\n",
    "#creo llista buida\n",
    "L1=[]\n",
    "#inicialitzo comptador a 0\n",
    "comp=0\n",
    "#mentres que la llargada de la llista sigui diferent al comptador\n",
    "while len(llista1)!=comp:\n",
    "    #poso que \"h\" és resi, la posició de la llista de aa, i or\n",
    "    h=['resi',llista1[comp],'or ']\n",
    "    #afegeixo h a la llista buida L1\n",
    "    L1.append(h)\n",
    "    #sumo 1 al comptador per canviar de posició\n",
    "    comp=comp+1\n",
    "#paso la llista a cadena\n",
    "s = ','.join(map(str, L1))\n",
    "#elimino: [ ] , '\n",
    "ss=s.replace('[','')\n",
    "sss=ss.replace(',','')\n",
    "ssss=sss.replace(']','')\n",
    "sssss=ssss.replace(\"'\",\"\")\n",
    "#ara borrem el \"or\" final \n",
    "Grin2B_up_aa = remove_last_or(sssss)\n",
    "Grin2B_up_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'resi 4 or resi 5 or resi 6 or resi 7 or resi 7 or resi 8 or resi 9 or resi 10 or resi 12 or resi 13 or resi 21 or resi 25 or resi 26 or resi 27 or resi 27 or resi 31 or resi 32 or resi 32 or resi 33 or resi 34 or resi 35 or resi 35 or resi 46 or resi 47 or resi 48 or resi 53 or resi 55 or resi 57 or resi 61 or resi 62 or resi 64 or resi 64 or resi 65 or resi 67 or resi 74 or resi 81 or resi 84 or resi 84 or resi 88 or resi 89 or resi 92 or resi 92 or resi 97 or resi 105 or resi 105 or resi 109 or resi 109 or resi 109 or resi 132 or resi 140 or resi 142 or resi 142 or resi 143 or resi 143 or resi 158 or resi 159 or resi 172 or resi 177 or resi 179 or resi 180 or resi 187 or resi 207 or resi 211 or resi 216 or resi 216 or resi 225 or resi 226 or resi 227 or resi 229 or resi 230 or resi 240 or resi 244 or resi 252 or resi 255 or resi 266 or resi 268 or resi 269 or resi 271 or resi 272 or resi 273 or resi 275 or resi 280 or resi 281 or resi 291 or resi 299 or resi 302 or resi 312 or resi 320 or resi 325 or resi 325 or resi 331 or resi 331 or resi 333 or resi 336 or resi 337 or resi 338 or resi 347 or resi 354 or resi 359 or resi 366 or resi 368 or resi 371 or resi 375 or resi 382 or resi 383 or resi 383 or resi 394 or resi 401 or resi 403 or resi 407 or resi 424 or resi 430 or resi 439 or resi 440 or resi 440 or resi 441 or resi 442 or resi 443 or resi 444 or resi 447 or resi 448 or resi 450 or resi 462 or resi 490 or resi 503 or resi 568 or resi 572 or resi 575 or resi 576 or resi 582 or resi 586 or resi 590 or resi 595 or resi 596 or resi 603 or resi 639 or resi 661 or resi 680 or resi 700 or resi 715 or resi 724 or resi 743 or resi 754 or resi 775 or resi 775 or resi 790 or resi 803 or resi 808 or resi 838 or resi 840 or resi 846 or resi 851 or resi 851 or resi 852 or resi 854 or resi 855 or resi 860 or resi 860 or resi 862 or resi 866 or resi 870 or resi 872 or resi 873 or resi 876 or resi 878 or resi 879 or resi 880 or resi 880 or resi 882 or resi 883 or resi 884 or resi 889 or resi 889 or resi 889 or resi 895 or resi 897 or resi 900 or resi 903 or resi 904 or resi 913 or resi 914 or resi 914 or resi 916 or resi 916 or resi 920 or resi 920 or resi 921 or resi 924 or resi 927 or resi 927 or resi 927 or resi 929 or resi 931 or resi 937 or resi 938 or resi 938 or resi 939 or resi 945 or resi 949 or resi 954 or resi 969 or resi 977 or resi 980 or resi 982 or resi 987 or resi 987 or resi 994 or resi 994 or resi 997 or resi 998 or resi 1002 or resi 1002 or resi 1005 or resi 1005 or resi 1010 or resi 1010 or resi 1013 or resi 1016 or resi 1017 or resi 1019 or resi 1023 or resi 1026 or resi 1031 or resi 1033 or resi 1035 or resi 1037 or resi 1040 or resi 1040 or resi 1044 or resi 1046 or resi 1048 or resi 1048 or resi 1048 or resi 1049 or resi 1055 or resi 1057 or resi 1063 or resi 1068 or resi 1075 or resi 1076 or resi 1078 or resi 1078 or resi 1079 or resi 1092 or resi 1099 or resi 1099 or resi 1100 or resi 1103 or resi 1110 or resi 1113 or resi 1113 or resi 1115 or resi 1115 or resi 1115 or resi 1115 or resi 1120 or resi 1121 or resi 1122 or resi 1122 or resi 1124 or resi 1128 or resi 1129 or resi 1130 or resi 1130 or resi 1136 or resi 1136 or resi 1136 or resi 1138 or resi 1138 or resi 1142 or resi 1145 or resi 1150 or resi 1153 or resi 1154 or resi 1155 or resi 1158 or resi 1158 or resi 1159 or resi 1159 or resi 1162 or resi 1164 or resi 1164 or resi 1164 or resi 1165 or resi 1166 or resi 1167 or resi 1168 or resi 1169 or resi 1172 or resi 1175 or resi 1178 or resi 1179 or resi 1179 or resi 1180 or resi 1181 or resi 1182 or resi 1188 or resi 1189 or resi 1189 or resi 1191 or resi 1192 or resi 1194 or resi 1195 or resi 1201 or resi 1203 or resi 1203 or resi 1209 or resi 1209 or resi 1211 or resi 1211 or resi 1213 or resi 1214 or resi 1216 or resi 1216 or resi 1217 or resi 1222 or resi 1228 or resi 1231 or resi 1234 or resi 1236 or resi 1241 or resi 1242 or resi 1250 or resi 1257 or resi 1266 or resi 1266 or resi 1267 or resi 1267 or resi 1269 or resi 1269 or resi 1271 or resi 1273 or resi 1275 or resi 1276 or resi 1276 or resi 1279 or resi 1279 or resi 1280 or resi 1282 or resi 1285 or resi 1291 or resi 1293 or resi 1295 or resi 1295 or resi 1299 or resi 1299 or resi 1305 or resi 1307 or resi 1308 or resi 1316 or resi 1320 or resi 1322 or resi 1327 or resi 1330 or resi 1331 or resi 1331 or resi 1331 or resi 1331 or resi 1332 or resi 1332 or resi 1332 or resi 1334 or resi 1338 or resi 1338 or resi 1339 or resi 1339 or resi 1344 or resi 1344 or resi 1344 or resi 1345 or resi 1346 or resi 1347 or resi 1348 or resi 1348 or resi 1350 or resi 1351 or resi 1354 or resi 1356 or resi 1357 or resi 1357 or resi 1359 or resi 1360 or resi 1368 or resi 1368 or resi 1369 or resi 1370 or resi 1371 or resi 1371 or resi 1372 or resi 1381 or resi 1383 or resi 1383 or resi 1384 or resi 1390 or resi 1401 or resi 1405 or resi 1410 or resi 1411 or resi 1412 or resi 1413 or resi 1414 or resi 1414 or resi 1414 or resi 1417 or resi 1417 or resi 1419 or resi 1419 or resi 1421 or resi 1422 or resi 1422 or resi 1427 or resi 1428 or resi 1429 or resi 1430 or resi 1430 or resi 1432 or resi 1434 or resi 1435 or resi 1438 or resi 1438 or resi 1439 or resi 1440 or resi 1440 or resi 1441 or resi 1441 or resi 1441 or resi 1441 or resi 1443 or resi 1444 or resi 1446 or resi 1447 or resi 1448 or resi 1449 or resi 1451 or resi 1451 or resi 1453 or resi 1457 or resi 1459 or resi 1460 or resi 1461 or resi 1461 or resi 1462 or resi 1462 or resi 1463 or resi 1464 or resi 1469'"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#GRIN 2B LLISTA Neutral\n",
    "#creo una llista amb la posició del aminoàcid\n",
    "llista1=list(G2B_n['Amino acid position'])\n",
    "#la ordeno per ordre creixent\n",
    "llista1.sort()\n",
    "#creo llista buida\n",
    "L1=[]\n",
    "#inicialitzo comptador a 0\n",
    "comp=0\n",
    "#mentres que la llargada de la llista sigui diferent al comptador\n",
    "while len(llista1)!=comp:\n",
    "    #poso que \"h\" és resi, la posició de la llista de aa, i or\n",
    "    h=['resi',llista1[comp],'or ']\n",
    "    #afegeixo h a la llista buida L1\n",
    "    L1.append(h)\n",
    "    #sumo 1 al comptador per canviar de posició\n",
    "    comp=comp+1\n",
    "#paso la llista a cadena\n",
    "s = ','.join(map(str, L1))\n",
    "#elimino: [ ] , '\n",
    "ss=s.replace('[','')\n",
    "sss=ss.replace(',','')\n",
    "ssss=sss.replace(']','')\n",
    "sssss=ssss.replace(\"'\",\"\")\n",
    "#ara borrem el \"or\" final \n",
    "#fem un replace de 'or' per 'res' a la ultima posició que trobi un or\n",
    "Grin2B_n_aa = remove_last_or(sssss)\n",
    "Grin2B_n_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "llista dels aa que estan a gnomAD V2 i no a GRIN db:  [['Gly', '1467', 'Ala'], ['Ser', '1452', 'Phe'], ['Pro', '1439', 'Ala'], ['Leu', '1424', 'Phe'], ['Ser', '1415', 'Leu'], ['Arg', '1329', 'Gln'], ['Val', '1322', 'Ile'], ['His', '1302', 'Gln'], ['Arg', '1300', 'Gln'], ['Val', '1272', 'Gly'], ['Asp', '1257', 'Ala'], ['Gly', '1212', 'Asp'], ['Asn', '1200', 'Lys'], ['His', '1187', 'Gln'], ['His', '1187', 'Gln'], ['His', '1187', 'Tyr'], ['Gly', '1184', 'Asp'], ['Gly', '1169', 'Ala'], ['Lys', '1156', 'Thr'], ['Arg', '1130', 'Trp'], ['Arg', '1121', 'His'], ['Arg', '1111', 'His'], ['Arg', '1057', 'His'], ['Arg', '990', 'Gln'], ['Gly', '971', 'Arg'], ['Pro', '952', 'Leu'], ['Arg', '926', 'Gln'], ['Arg', '900', 'Cys'], ['Trp', '844', 'Arg'], ['Ala', '827', 'Val'], ['Val', '821', 'Gly'], ['Gly', '820', 'Ala'], ['Glu', '807', 'Val'], ['His', '802', 'Gln'], ['Arg', '371', 'Lys'], ['Gly', '346', 'Val'], ['Ala', '291', 'Val'], ['Ala', '271', 'Val'], ['Ile', '168', 'Thr'], ['Met', '134', 'Ile'], ['Asp', '104', 'Glu'], ['Arg', '84', 'Ser'], ['Arg', '27', 'Cys'], ['Val', '18', 'Ile'], ['Ala', '17', 'Thr'], ['Ala', '5', 'Thr']]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ys/tvdvp6kj24df4dh9qb_87_tr0000gn/T/ipykernel_43013/1681877237.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2B=G2B_rev[['lletres', 'numeros']] = G2B_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n",
      "/var/folders/ys/tvdvp6kj24df4dh9qb_87_tr0000gn/T/ipykernel_43013/1681877237.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2B=G2B_rev[['lletres', 'numeros']] = G2B_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n"
     ]
    }
   ],
   "source": [
    "#COMPARACIÓ GRIN2B GRINDB VS GNOMAD V2\n",
    "\n",
    "#De les que tinc a la base de dades\n",
    "#Per fer Met654Ile de les neutral\n",
    "#copio la llista de grin1 neutral\n",
    "G2B_aa=G2B_n\n",
    "#alimino totes les columnes que no necessito\n",
    "G2B_aa=G2B_aa.drop(['Gene','Gain/Loss','Disease-association','Mutation Category'], axis=1)\n",
    "#les reordeno\n",
    "G2B_aa = G2B_aa.reindex(columns=['Reference Amino acid','Amino acid position','Mutated Amino acid'])\n",
    "#Poso com string la posició del aa mutat\n",
    "G2B_aa['string_numbers'] = G2B_aa['Amino acid position'].astype(str)\n",
    "#alimino totes les columnes que no necessito\n",
    "G2B_aa=G2B_aa.drop(['Amino acid position'], axis=1)\n",
    "#les reordeno\n",
    "G2B_aa = G2B_aa.reindex(columns=['Reference Amino acid','string_numbers','Mutated Amino acid'])\n",
    "#ho paso a llista\n",
    "G2B_aaL=G2B_aa.to_numpy().tolist()\n",
    "#ara tinc la llista dels aa mutació neutra de grin1 que estan a la base de dades\n",
    "G2B_aaL\n",
    "\n",
    "\n",
    "#llegeixo arxiu gnomad v2\n",
    "dades = pd.read_csv(\"/Users/ismaelvera/Desktop/TFG/Mutacions GRINdb/GRIN_missense_variants_gnomad_v2_1.txt\",header=None) \n",
    "#em quedo nomes amb les columnes que diuen la mutació i el gen\n",
    "dades=dades.drop([0,1,2,3,4,5,6,8,10,11,12,13,14], axis=1)\n",
    "#faig una llista amb els que son nomes de grin1\n",
    "G2B_rev=dades.loc[dades[9]=='GRIN2B']\n",
    "#com que ho tinc com a p.Ser2Arg\n",
    "#creo funció per separar lletres de numeros \n",
    "def separate_letters_numbers(text):\n",
    "    lletres = ''.join(filter(str.isalpha, text))\n",
    "    numeros = ''.join(filter(str.isdigit, text))\n",
    "    return (lletres, numeros)\n",
    "#aplico funció a la columna que m'interesa\n",
    "df2B=G2B_rev[['lletres', 'numeros']] = G2B_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n",
    "#ara tinc lletres i numeros separats\n",
    "#separo les lletres per tenir p, Ser, Arg\n",
    "df2B['col1'] = df2B[0].str[0]\n",
    "df2B['col2'] = df2B[0].str[1:4]\n",
    "df2B['col3'] = df2B[0].str[4:]\n",
    "#elimino totes les columnes que no necessito\n",
    "df2B=df2B.drop([0,'col1'], axis=1)\n",
    "#les reordeno\n",
    "df2B = df2B.reindex(columns=['col2',1,'col3'])\n",
    "#ho paso a llista\n",
    "df2L2B=df2B.to_numpy().tolist()\n",
    "\n",
    "\n",
    "#faig la comparació\n",
    "#he de crear una llista buida on guardare els que no estan \n",
    "noestanG2BV2=[]\n",
    "#per cada element que tinc a la llista de gnomad v2, si no esta a la llista dels \n",
    "# que tinc a la base de dades, mel afegeix a la llista L1\n",
    "for element in df2L2B:\n",
    "    if element not in G2B_aaL:\n",
    "        noestanG2BV2.append(element)\n",
    "\n",
    "\n",
    "#ara per poder marcar-los al model molecular he de fer que posi resi 23 or resi 234 ....\n",
    "#primer separo la llista L1, que son els que no tinc en AA1 posició AA2\n",
    "df2B_2 = pd.DataFrame(noestanG2BV2, columns=['AA 1', 'Position', 'AA 2'])\n",
    "#creo una llista amb la posició del aminoàcid\n",
    "llista1=list(df2B_2['Position'])\n",
    "#la ordeno per ordre creixent\n",
    "llista1.sort()\n",
    "#creo llista buida\n",
    "L1=[]\n",
    "#inicialitzo comptador a 0\n",
    "comp=0\n",
    "#mentres que la llargada de la llista sigui diferent al comptador\n",
    "while len(llista1)!=comp:\n",
    "    #poso que \"h\" és resi, la posició de la llista de aa, i or\n",
    "    h=['resi',llista1[comp],'or ']\n",
    "    #afegeixo h a la llista buida L1\n",
    "    L1.append(h)\n",
    "    #sumo 1 al comptador per canviar de posició\n",
    "    comp=comp+1\n",
    "#paso la llista a cadena\n",
    "s = ','.join(map(str, L1))\n",
    "#elimino: [ ] , '\n",
    "ss=s.replace('[','')\n",
    "sss=ss.replace(',','')\n",
    "ssss=sss.replace(']','')\n",
    "sssss=ssss.replace(\"'\",\"\")\n",
    "#ara borrem el \"or\" final \n",
    "#faig una funció per borrar el or final\n",
    "def remove_last_or(chain):\n",
    "    #Divideix la cadena en diferents elements separats per \"or\"\n",
    "    elements = chain.split(\"or\")\n",
    "    #Per si nomes hagues un a la llista, retorna la cadena tal com està\n",
    "    if len(elements) == 1:\n",
    "        return chain\n",
    "    #Treu qualsevol espai en blanc inicial o final de cada element\n",
    "    cleaned_elements = [elem.strip() for elem in elements]\n",
    "    #Elimina l'últim element, que hauria de ser una cadena buida després de l'últim \"or\"\n",
    "    cleaned_elements.pop()\n",
    "    #Torna a unir els elements \"netejats\" fent servir \"or\" com a separador\n",
    "    new_chain = \" or \".join(cleaned_elements)\n",
    "    return new_chain\n",
    "G2B_rev_v2 = remove_last_or(sssss)\n",
    "G2B_rev_v2\n",
    "\n",
    "#mostro la llista dels que no estan\n",
    "print('llista dels aa que estan a gnomAD V2 i no a GRIN db: ',noestanG2BV2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "llista dels aa que estan a gnomAD V2 i no a GRIN db:  [['Pro', '44', 'Leu'], ['Pro', '44', 'Thr'], ['Val', '41', 'Ala'], ['Lys', '36', 'Glu'], ['Ile', '33', 'Thr'], ['Arg', '31', 'Ser'], ['Leu', '27', 'Gln'], ['Met', '24', 'Thr'], ['Leu', '1477', 'Val'], ['Gly', '1467', 'Ala'], ['Gly', '1467', 'Ser'], ['Pro', '1462', 'Leu'], ['Pro', '1454', 'Arg'], ['Pro', '1454', 'Ser'], ['Asn', '1453', 'Lys'], ['Ile', '1448', 'Val'], ['Ile', '1446', 'Thr'], ['Gly', '1436', 'Arg'], ['Leu', '1424', 'Phe'], ['Ala', '1417', 'Ser'], ['Ala', '1412', 'Glu'], ['Lys', '1402', 'Arg'], ['Ser', '1401', 'Gly'], ['Arg', '1381', 'Trp'], ['Pro', '1367', 'Leu'], ['Asn', '1365', 'His'], ['Ala', '1350', 'Ser'], ['Glu', '1346', 'Asp'], ['Glu', '1341', 'Lys'], ['Val', '1322', 'Ile'], ['Arg', '1300', 'Gln'], ['Arg', '1300', 'Trp'], ['Lys', '1297', 'Thr'], ['Thr', '1279', 'Ile'], ['Asn', '1250', 'Lys'], ['Ala', '1248', 'Val'], ['Val', '1229', 'Met'], ['Lys', '1221', 'Arg'], ['Cys', '1218', 'Ser'], ['Cys', '1215', 'Gly'], ['Ser', '1210', 'Pro'], ['Trp', '1206', 'Arg'], ['Pro', '1194', 'Arg'], ['Val', '1189', 'Gly'], ['Lys', '1186', 'Asn'], ['Asp', '1185', 'Glu'], ['Thr', '1183', 'Pro'], ['Ile', '1179', 'Phe'], ['His', '1178', 'Tyr'], ['Asn', '1175', 'Lys'], ['Ser', '1166', 'Cys'], ['Lys', '1156', 'Met'], ['Lys', '1156', 'Thr'], ['Asp', '1153', 'Glu'], ['Asp', '1153', 'Glu'], ['Asp', '1150', 'Glu'], ['Val', '1149', 'Gly'], ['Glu', '1147', 'Asp'], ['His', '1145', 'Gln'], ['Ser', '1143', 'Thr'], ['Arg', '1138', 'Gly'], ['Arg', '1121', 'His'], ['His', '1119', 'Arg'], ['Lys', '1082', 'Arg'], ['Arg', '1080', 'Lys'], ['Arg', '1057', 'His'], ['Asp', '1054', 'His'], ['Ser', '1050', 'Gly'], ['Gln', '1034', 'Glu'], ['His', '1032', 'Gln'], ['Pro', '1028', 'Ser'], ['Lys', '1021', 'Glu'], ['Ser', '1017', 'Ala'], ['Ser', '1015', 'Ala'], ['Asp', '1001', 'His'], ['Arg', '990', 'Gln'], ['Val', '980', 'Gly'], ['Asn', '979', 'Ile'], ['Gly', '971', 'Arg'], ['Asp', '961', 'Asn'], ['Tyr', '932', 'Cys'], ['Ile', '925', 'Val'], ['Ala', '921', 'Pro'], ['Thr', '894', 'Ile'], ['Gly', '874', 'Glu'], ['Gly', '859', 'Ala'], ['Arg', '847', 'Gln'], ['Trp', '844', 'Arg'], ['Ile', '837', 'Phe'], ['Leu', '833', 'Val'], ['Ala', '827', 'Val'], ['Ala', '827', 'Glu'], ['Gly', '787', 'Ala'], ['Gly', '713', 'Ala'], ['Arg', '682', 'His'], ['Pro', '680', 'Leu'], ['Val', '618', 'Ile'], ['Phe', '580', 'Tyr'], ['Ser', '520', 'Leu'], ['Glu', '518', 'Asp'], ['Val', '441', 'Ile'], ['Arg', '439', 'Leu'], ['Pro', '424', 'Thr'], ['Lys', '382', 'Asn'], ['Arg', '371', 'Gly'], ['Lys', '369', 'Arg'], ['Val', '363', 'Leu'], ['Gly', '346', 'Arg'], ['Glu', '345', 'Ala'], ['Ile', '329', 'Met'], ['Pro', '274', 'Leu'], ['Ala', '271', 'Val'], ['Ala', '271', 'Ser'], ['Pro', '259', 'His'], ['Tyr', '254', 'Cys'], ['Thr', '238', 'Ala'], ['Glu', '200', 'Gln'], ['Ile', '190', 'Val'], ['Arg', '187', 'Cys'], ['Ile', '171', 'Val'], ['Ser', '149', 'Pro'], ['Gly', '96', 'Arg'], ['Val', '68', 'Met'], ['Val', '42', 'Ala'], ['Val', '39', 'Ala'], ['Pro', '32', 'Ala'], ['Ser', '24', 'Asn'], ['Val', '18', 'Ile']]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/ys/tvdvp6kj24df4dh9qb_87_tr0000gn/T/ipykernel_43013/3458456916.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2B=G2B_rev[['lletres', 'numeros']] = G2B_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n",
      "/var/folders/ys/tvdvp6kj24df4dh9qb_87_tr0000gn/T/ipykernel_43013/3458456916.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df2B=G2B_rev[['lletres', 'numeros']] = G2B_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n"
     ]
    }
   ],
   "source": [
    "#COMPARACIÓ GRIN2B GRINDB VS GNOMAD V3\n",
    "\n",
    "#De les que tinc a la base de dades\n",
    "#Per fer Met654Ile de les neutral\n",
    "#copio la llista de grin1 neutral\n",
    "G2B_aa=G2B_n\n",
    "#alimino totes les columnes que no necessito\n",
    "G2B_aa=G2B_aa.drop(['Gene','Gain/Loss','Disease-association','Mutation Category'], axis=1)\n",
    "#les reordeno\n",
    "G2B_aa = G2B_aa.reindex(columns=['Reference Amino acid','Amino acid position','Mutated Amino acid'])\n",
    "#Poso com string la posició del aa mutat\n",
    "G2B_aa['string_numbers'] = G2B_aa['Amino acid position'].astype(str)\n",
    "#alimino totes les columnes que no necessito\n",
    "G2B_aa=G2B_aa.drop(['Amino acid position'], axis=1)\n",
    "#les reordeno\n",
    "G2B_aa = G2B_aa.reindex(columns=['Reference Amino acid','string_numbers','Mutated Amino acid'])\n",
    "#ho paso a llista\n",
    "G2B_aaL=G2B_aa.to_numpy().tolist()\n",
    "#ara tinc la llista dels aa mutació neutra de grin1 que estan a la base de dades\n",
    "G2B_aaL\n",
    "\n",
    "\n",
    "#llegeixo arxiu gnomad v3\n",
    "dades = pd.read_csv(\"/Users/ismaelvera/Desktop/TFG/Mutacions GRINdb/GRIN_missense_variants_gnomad_v3.txt\",header=None) \n",
    "#em quedo nomes amb les columnes que diuen la mutació i el gen\n",
    "dades=dades.drop([0,1,2,3,4,5,6,8,10,11,12,13,14], axis=1)\n",
    "#faig una llista amb els que son nomes de grin1\n",
    "G2B_rev=dades.loc[dades[9]=='GRIN2B']\n",
    "#com que ho tinc com a p.Ser2Arg\n",
    "#creo funció per separar lletres de numeros \n",
    "def separate_letters_numbers(text):\n",
    "    lletres = ''.join(filter(str.isalpha, text))\n",
    "    numeros = ''.join(filter(str.isdigit, text))\n",
    "    return (lletres, numeros)\n",
    "#aplico funció a la columna que m'interesa\n",
    "df2B=G2B_rev[['lletres', 'numeros']] = G2B_rev[7].apply(separate_letters_numbers).apply(pd.Series)\n",
    "#ara tinc lletres i numeros separats\n",
    "#separo les lletres per tenir p, Ser, Arg\n",
    "df2B['col1'] = df2B[0].str[0]\n",
    "df2B['col2'] = df2B[0].str[1:4]\n",
    "df2B['col3'] = df2B[0].str[4:]\n",
    "#elimino totes les columnes que no necessito\n",
    "df2B=df2B.drop([0,'col1'], axis=1)\n",
    "#les reordeno\n",
    "df2B = df2B.reindex(columns=['col2',1,'col3'])\n",
    "#ho paso a llista\n",
    "df2L2B=df2B.to_numpy().tolist()\n",
    "\n",
    "\n",
    "#faig la comparació\n",
    "#he de crear una llista buida on guardare els que no estan \n",
    "noestanG2BV3=[]\n",
    "#per cada element que tinc a la llista de gnomad v2, si no esta a la llista dels \n",
    "# que tinc a la base de dades, mel afegeix a la llista L1\n",
    "for element in df2L2B:\n",
    "    if element not in G2B_aaL:\n",
    "        noestanG2BV3.append(element)\n",
    "\n",
    "\n",
    "#ara per poder marcar-los al model molecular he de fer que posi resi 23 or resi 234 ....\n",
    "#primer separo la llista L1, que son els que no tinc en AA1 posició AA2\n",
    "df2B_2 = pd.DataFrame(noestanG2BV3, columns=['AA 1', 'Position', 'AA 2'])\n",
    "#creo una llista amb la posició del aminoàcid\n",
    "llista1=list(df2B_2['Position'])\n",
    "#la ordeno per ordre creixent\n",
    "llista1.sort()\n",
    "#creo llista buida\n",
    "L1=[]\n",
    "#inicialitzo comptador a 0\n",
    "comp=0\n",
    "#mentres que la llargada de la llista sigui diferent al comptador\n",
    "while len(llista1)!=comp:\n",
    "    #poso que \"h\" és resi, la posició de la llista de aa, i or\n",
    "    h=['resi',llista1[comp],'or ']\n",
    "    #afegeixo h a la llista buida L1\n",
    "    L1.append(h)\n",
    "    #sumo 1 al comptador per canviar de posició\n",
    "    comp=comp+1\n",
    "#paso la llista a cadena\n",
    "s = ','.join(map(str, L1))\n",
    "#elimino: [ ] , '\n",
    "ss=s.replace('[','')\n",
    "sss=ss.replace(',','')\n",
    "ssss=sss.replace(']','')\n",
    "sssss=ssss.replace(\"'\",\"\")\n",
    "#ara borrem el \"or\" final \n",
    "#faig una funció per borrar el or final\n",
    "def remove_last_or(chain):\n",
    "    #Divideix la cadena en diferents elements separats per \"or\"\n",
    "    elements = chain.split(\"or\")\n",
    "    #Per si nomes hagues un a la llista, retorna la cadena tal com està\n",
    "    if len(elements) == 1:\n",
    "        return chain\n",
    "    #Treu qualsevol espai en blanc inicial o final de cada element\n",
    "    cleaned_elements = [elem.strip() for elem in elements]\n",
    "    #Elimina l'últim element, que hauria de ser una cadena buida després de l'últim \"or\"\n",
    "    cleaned_elements.pop()\n",
    "    #Torna a unir els elements \"netejats\" fent servir \"or\" com a separador\n",
    "    new_chain = \" or \".join(cleaned_elements)\n",
    "    return new_chain\n",
    "G2B_rev_v3 = remove_last_or(sssss)\n",
    "G2B_rev_v3\n",
    "\n",
    "#mostro la llista dels que no estan\n",
    "print('llista dels aa que estan a gnomAD V2 i no a GRIN db: ',noestanG2BV3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Escribim fitxer de text i marquem amb colors els aa mutats\n",
    "#Primer carreguem el programa i pintem amb diferents colors les cadenes\n",
    "#Disease association: roig\n",
    "#Uncertain pathogenesis: taronja\n",
    "#Neutral: verd\n",
    "text=open(\"Patogenecitat i revisio.txt\",\"w\") \n",
    "text.write(\"load /Users/ismaelvera/Desktop/TFG/Mutacions GRINdb/GRIN.pdb\\n\"\n",
    "           \"hide everything\\n\"\n",
    "           \"show cartoon\\n\"\n",
    "           \"sele chain A+C\\n\"\n",
    "           \" color cyan, sele\\n\"\n",
    "           \"sele chain B\\n\"\n",
    "           \" color lightblue,sele\\n\"\n",
    "           \"sele chain D\\n\"\n",
    "           \" color marine,sele\\n\") \n",
    "text.write('sele (chain A+C and name ca and (')\n",
    "text.write(Grin1_da_aa)\n",
    "text.write(\"))\\n\"\n",
    "           \"show spheres, sele\\n\"\n",
    "           \"color red, sele\\n\")\n",
    "text.write('sele (chain A+C and name ca and (')\n",
    "text.write(Grin1_up_aa)\n",
    "text.write(\"))\\n\"\n",
    "           \"show spheres, sele\\n\"\n",
    "           \"color orange, sele\\n\")\n",
    "text.write('sele (chain A+C and name ca and (')\n",
    "text.write(Grin1_n_aa)\n",
    "text.write(\"))\\n\"\n",
    "           \"show spheres, sele\\n\"\n",
    "           \"color green, sele\\n\")\n",
    "text.write('sele (chain B and name ca and (')\n",
    "text.write(Grin2A_da_aa)\n",
    "text.write(\"))\\n\"\n",
    "           \"show spheres, sele\\n\"\n",
    "           \"color red, sele\\n\")\n",
    "text.write('sele (chain B and name ca and (')\n",
    "text.write(Grin2A_up_aa)\n",
    "text.write(\"))\\n\"\n",
    "           \"show spheres, sele\\n\"\n",
    "           \"color orange, sele\\n\")\n",
    "text.write('sele (chain B and name ca and (')\n",
    "text.write(Grin2A_n_aa)\n",
    "text.write(\"))\\n\"\n",
    "           \"show spheres, sele\\n\"\n",
    "           \"color green, sele\\n\")\n",
    "text.write('sele (chain D and name ca and (')\n",
    "text.write(Grin2B_da_aa)\n",
    "text.write(\"))\\n\"\n",
    "           \"show spheres, sele\\n\"\n",
    "           \"color red, sele\\n\")\n",
    "text.write('sele (chain D and name ca and (')\n",
    "text.write(Grin2B_up_aa)\n",
    "text.write(\"))\\n\"\n",
    "           \"show spheres, sele\\n\"\n",
    "           \"color orange, sele\\n\")\n",
    "text.write('sele (chain D and name ca and (')\n",
    "text.write(Grin2B_n_aa)\n",
    "text.write(\"))\\n\"\n",
    "           \"show spheres, sele\\n\"\n",
    "           \"color green, sele\\n\")\n",
    "text.write('sele (chain A+C and name ca and (')\n",
    "text.write(G1_rev_v2)\n",
    "text.write(\"))\\n\"\n",
    "           \"show spheres, sele\\n\"\n",
    "           \"color green, sele\\n\")\n",
    "text.write('sele (chain A+C and name ca and (')\n",
    "text.write(G1_rev_v3)\n",
    "text.write(\"))\\n\"\n",
    "           \"show spheres, sele\\n\"\n",
    "           \"color green, sele\\n\")\n",
    "text.write('sele (chain B and name ca and (')\n",
    "text.write(G2A_rev_v2)\n",
    "text.write(\"))\\n\"\n",
    "           \"show spheres, sele\\n\"\n",
    "           \"color green, sele\\n\")\n",
    "text.write('sele (chain B and name ca and (')\n",
    "text.write(G2A_rev_v3)\n",
    "text.write(\"))\\n\"\n",
    "           \"show spheres, sele\\n\"\n",
    "           \"color green, sele\\n\")\n",
    "text.write('sele (chain D and name ca and (')\n",
    "text.write(G2B_rev_v2)\n",
    "text.write(\"))\\n\"\n",
    "           \"show spheres, sele\\n\"\n",
    "           \"color green, sele\\n\")\n",
    "text.write('sele (chain D and name ca and (')\n",
    "text.write(G2B_rev_v3)\n",
    "text.write(\"))\\n\"\n",
    "           \"show spheres, sele\\n\"\n",
    "           \"color green, sele\\n\")\n",
    "text.close() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Patogenecitat i revisio 2.txt'"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Faig una copia del arxiu que he creat\n",
    "import shutil\n",
    "shutil.copy('Patogenecitat i revisio.txt', 'Patogenecitat i revisio 2.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Canvio el nom de la copia per fer-ho pml, aixi ho tinc en txt i en pml\n",
    "import os\n",
    "os.rename('Patogenecitat i revisio 2.txt', 'Patogenecitat i revisio.pml')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "vscode": {
   "interpreter": {
    "hash": "0d6fd54eaddb41f6436496db92bf07f7f19b9b9d19e6b38ccb3c795494a23e8b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
